{
  "nbformat": 4,
  "nbformat_minor": 0,
  "metadata": {
    "colab": {
      "provenance": []
    },
    "kernelspec": {
      "name": "python3",
      "display_name": "Python 3"
    },
    "language_info": {
      "name": "python"
    }
  },
  "cells": [
    {
      "cell_type": "code",
      "source": [
        "! pip install spacy\n",
        "! python -m spacy download en_core_web_sm\n",
        "! pip install top2vec[sentence_transformers]\n",
        "! pip install top2vec[sentence_encoders]\n",
        "! pip install tensorflow tensorflow_hub tensorflow_text\n",
        "! pip install colored"
      ],
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "KZozv84mGxfs",
        "outputId": "c4858fad-9ace-4fb7-c48d-d0011b6bc350"
      },
      "execution_count": 1,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "Requirement already satisfied: spacy in /usr/local/lib/python3.10/dist-packages (3.7.2)\n",
            "Requirement already satisfied: spacy-legacy<3.1.0,>=3.0.11 in /usr/local/lib/python3.10/dist-packages (from spacy) (3.0.12)\n",
            "Requirement already satisfied: spacy-loggers<2.0.0,>=1.0.0 in /usr/local/lib/python3.10/dist-packages (from spacy) (1.0.5)\n",
            "Requirement already satisfied: murmurhash<1.1.0,>=0.28.0 in /usr/local/lib/python3.10/dist-packages (from spacy) (1.0.10)\n",
            "Requirement already satisfied: cymem<2.1.0,>=2.0.2 in /usr/local/lib/python3.10/dist-packages (from spacy) (2.0.8)\n",
            "Requirement already satisfied: preshed<3.1.0,>=3.0.2 in /usr/local/lib/python3.10/dist-packages (from spacy) (3.0.9)\n",
            "Requirement already satisfied: thinc<8.3.0,>=8.1.8 in /usr/local/lib/python3.10/dist-packages (from spacy) (8.2.3)\n",
            "Requirement already satisfied: wasabi<1.2.0,>=0.9.1 in /usr/local/lib/python3.10/dist-packages (from spacy) (1.1.2)\n",
            "Requirement already satisfied: srsly<3.0.0,>=2.4.3 in /usr/local/lib/python3.10/dist-packages (from spacy) (2.4.8)\n",
            "Requirement already satisfied: catalogue<2.1.0,>=2.0.6 in /usr/local/lib/python3.10/dist-packages (from spacy) (2.0.10)\n",
            "Requirement already satisfied: weasel<0.4.0,>=0.1.0 in /usr/local/lib/python3.10/dist-packages (from spacy) (0.3.4)\n",
            "Requirement already satisfied: typer<0.10.0,>=0.3.0 in /usr/local/lib/python3.10/dist-packages (from spacy) (0.9.0)\n",
            "Requirement already satisfied: smart-open<7.0.0,>=5.2.1 in /usr/local/lib/python3.10/dist-packages (from spacy) (6.4.0)\n",
            "Requirement already satisfied: tqdm<5.0.0,>=4.38.0 in /usr/local/lib/python3.10/dist-packages (from spacy) (4.66.2)\n",
            "Requirement already satisfied: requests<3.0.0,>=2.13.0 in /usr/local/lib/python3.10/dist-packages (from spacy) (2.31.0)\n",
            "Requirement already satisfied: pydantic!=1.8,!=1.8.1,<3.0.0,>=1.7.4 in /usr/local/lib/python3.10/dist-packages (from spacy) (2.6.1)\n",
            "Requirement already satisfied: jinja2 in /usr/local/lib/python3.10/dist-packages (from spacy) (3.1.3)\n",
            "Requirement already satisfied: setuptools in /usr/local/lib/python3.10/dist-packages (from spacy) (67.7.2)\n",
            "Requirement already satisfied: packaging>=20.0 in /usr/local/lib/python3.10/dist-packages (from spacy) (23.2)\n",
            "Requirement already satisfied: langcodes<4.0.0,>=3.2.0 in /usr/local/lib/python3.10/dist-packages (from spacy) (3.3.0)\n",
            "Requirement already satisfied: numpy>=1.19.0 in /usr/local/lib/python3.10/dist-packages (from spacy) (1.25.2)\n",
            "Requirement already satisfied: annotated-types>=0.4.0 in /usr/local/lib/python3.10/dist-packages (from pydantic!=1.8,!=1.8.1,<3.0.0,>=1.7.4->spacy) (0.6.0)\n",
            "Requirement already satisfied: pydantic-core==2.16.2 in /usr/local/lib/python3.10/dist-packages (from pydantic!=1.8,!=1.8.1,<3.0.0,>=1.7.4->spacy) (2.16.2)\n",
            "Requirement already satisfied: typing-extensions>=4.6.1 in /usr/local/lib/python3.10/dist-packages (from pydantic!=1.8,!=1.8.1,<3.0.0,>=1.7.4->spacy) (4.9.0)\n",
            "Requirement already satisfied: charset-normalizer<4,>=2 in /usr/local/lib/python3.10/dist-packages (from requests<3.0.0,>=2.13.0->spacy) (3.3.2)\n",
            "Requirement already satisfied: idna<4,>=2.5 in /usr/local/lib/python3.10/dist-packages (from requests<3.0.0,>=2.13.0->spacy) (3.6)\n",
            "Requirement already satisfied: urllib3<3,>=1.21.1 in /usr/local/lib/python3.10/dist-packages (from requests<3.0.0,>=2.13.0->spacy) (2.0.7)\n",
            "Requirement already satisfied: certifi>=2017.4.17 in /usr/local/lib/python3.10/dist-packages (from requests<3.0.0,>=2.13.0->spacy) (2024.2.2)\n",
            "Requirement already satisfied: blis<0.8.0,>=0.7.8 in /usr/local/lib/python3.10/dist-packages (from thinc<8.3.0,>=8.1.8->spacy) (0.7.11)\n",
            "Requirement already satisfied: confection<1.0.0,>=0.0.1 in /usr/local/lib/python3.10/dist-packages (from thinc<8.3.0,>=8.1.8->spacy) (0.1.4)\n",
            "Requirement already satisfied: click<9.0.0,>=7.1.1 in /usr/local/lib/python3.10/dist-packages (from typer<0.10.0,>=0.3.0->spacy) (8.1.7)\n",
            "Requirement already satisfied: cloudpathlib<0.17.0,>=0.7.0 in /usr/local/lib/python3.10/dist-packages (from weasel<0.4.0,>=0.1.0->spacy) (0.16.0)\n",
            "Requirement already satisfied: MarkupSafe>=2.0 in /usr/local/lib/python3.10/dist-packages (from jinja2->spacy) (2.1.5)\n",
            "Collecting en-core-web-sm==3.7.1\n",
            "  Downloading https://github.com/explosion/spacy-models/releases/download/en_core_web_sm-3.7.1/en_core_web_sm-3.7.1-py3-none-any.whl (12.8 MB)\n",
            "\u001b[2K     \u001b[90m━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━\u001b[0m \u001b[32m12.8/12.8 MB\u001b[0m \u001b[31m63.6 MB/s\u001b[0m eta \u001b[36m0:00:00\u001b[0m\n",
            "\u001b[?25hRequirement already satisfied: spacy<3.8.0,>=3.7.2 in /usr/local/lib/python3.10/dist-packages (from en-core-web-sm==3.7.1) (3.7.2)\n",
            "Requirement already satisfied: spacy-legacy<3.1.0,>=3.0.11 in /usr/local/lib/python3.10/dist-packages (from spacy<3.8.0,>=3.7.2->en-core-web-sm==3.7.1) (3.0.12)\n",
            "Requirement already satisfied: spacy-loggers<2.0.0,>=1.0.0 in /usr/local/lib/python3.10/dist-packages (from spacy<3.8.0,>=3.7.2->en-core-web-sm==3.7.1) (1.0.5)\n",
            "Requirement already satisfied: murmurhash<1.1.0,>=0.28.0 in /usr/local/lib/python3.10/dist-packages (from spacy<3.8.0,>=3.7.2->en-core-web-sm==3.7.1) (1.0.10)\n",
            "Requirement already satisfied: cymem<2.1.0,>=2.0.2 in /usr/local/lib/python3.10/dist-packages (from spacy<3.8.0,>=3.7.2->en-core-web-sm==3.7.1) (2.0.8)\n",
            "Requirement already satisfied: preshed<3.1.0,>=3.0.2 in /usr/local/lib/python3.10/dist-packages (from spacy<3.8.0,>=3.7.2->en-core-web-sm==3.7.1) (3.0.9)\n",
            "Requirement already satisfied: thinc<8.3.0,>=8.1.8 in /usr/local/lib/python3.10/dist-packages (from spacy<3.8.0,>=3.7.2->en-core-web-sm==3.7.1) (8.2.3)\n",
            "Requirement already satisfied: wasabi<1.2.0,>=0.9.1 in /usr/local/lib/python3.10/dist-packages (from spacy<3.8.0,>=3.7.2->en-core-web-sm==3.7.1) (1.1.2)\n",
            "Requirement already satisfied: srsly<3.0.0,>=2.4.3 in /usr/local/lib/python3.10/dist-packages (from spacy<3.8.0,>=3.7.2->en-core-web-sm==3.7.1) (2.4.8)\n",
            "Requirement already satisfied: catalogue<2.1.0,>=2.0.6 in /usr/local/lib/python3.10/dist-packages (from spacy<3.8.0,>=3.7.2->en-core-web-sm==3.7.1) (2.0.10)\n",
            "Requirement already satisfied: weasel<0.4.0,>=0.1.0 in /usr/local/lib/python3.10/dist-packages (from spacy<3.8.0,>=3.7.2->en-core-web-sm==3.7.1) (0.3.4)\n",
            "Requirement already satisfied: typer<0.10.0,>=0.3.0 in /usr/local/lib/python3.10/dist-packages (from spacy<3.8.0,>=3.7.2->en-core-web-sm==3.7.1) (0.9.0)\n",
            "Requirement already satisfied: smart-open<7.0.0,>=5.2.1 in /usr/local/lib/python3.10/dist-packages (from spacy<3.8.0,>=3.7.2->en-core-web-sm==3.7.1) (6.4.0)\n",
            "Requirement already satisfied: tqdm<5.0.0,>=4.38.0 in /usr/local/lib/python3.10/dist-packages (from spacy<3.8.0,>=3.7.2->en-core-web-sm==3.7.1) (4.66.2)\n",
            "Requirement already satisfied: requests<3.0.0,>=2.13.0 in /usr/local/lib/python3.10/dist-packages (from spacy<3.8.0,>=3.7.2->en-core-web-sm==3.7.1) (2.31.0)\n",
            "Requirement already satisfied: pydantic!=1.8,!=1.8.1,<3.0.0,>=1.7.4 in /usr/local/lib/python3.10/dist-packages (from spacy<3.8.0,>=3.7.2->en-core-web-sm==3.7.1) (2.6.1)\n",
            "Requirement already satisfied: jinja2 in /usr/local/lib/python3.10/dist-packages (from spacy<3.8.0,>=3.7.2->en-core-web-sm==3.7.1) (3.1.3)\n",
            "Requirement already satisfied: setuptools in /usr/local/lib/python3.10/dist-packages (from spacy<3.8.0,>=3.7.2->en-core-web-sm==3.7.1) (67.7.2)\n",
            "Requirement already satisfied: packaging>=20.0 in /usr/local/lib/python3.10/dist-packages (from spacy<3.8.0,>=3.7.2->en-core-web-sm==3.7.1) (23.2)\n",
            "Requirement already satisfied: langcodes<4.0.0,>=3.2.0 in /usr/local/lib/python3.10/dist-packages (from spacy<3.8.0,>=3.7.2->en-core-web-sm==3.7.1) (3.3.0)\n",
            "Requirement already satisfied: numpy>=1.19.0 in /usr/local/lib/python3.10/dist-packages (from spacy<3.8.0,>=3.7.2->en-core-web-sm==3.7.1) (1.25.2)\n",
            "Requirement already satisfied: annotated-types>=0.4.0 in /usr/local/lib/python3.10/dist-packages (from pydantic!=1.8,!=1.8.1,<3.0.0,>=1.7.4->spacy<3.8.0,>=3.7.2->en-core-web-sm==3.7.1) (0.6.0)\n",
            "Requirement already satisfied: pydantic-core==2.16.2 in /usr/local/lib/python3.10/dist-packages (from pydantic!=1.8,!=1.8.1,<3.0.0,>=1.7.4->spacy<3.8.0,>=3.7.2->en-core-web-sm==3.7.1) (2.16.2)\n",
            "Requirement already satisfied: typing-extensions>=4.6.1 in /usr/local/lib/python3.10/dist-packages (from pydantic!=1.8,!=1.8.1,<3.0.0,>=1.7.4->spacy<3.8.0,>=3.7.2->en-core-web-sm==3.7.1) (4.9.0)\n",
            "Requirement already satisfied: charset-normalizer<4,>=2 in /usr/local/lib/python3.10/dist-packages (from requests<3.0.0,>=2.13.0->spacy<3.8.0,>=3.7.2->en-core-web-sm==3.7.1) (3.3.2)\n",
            "Requirement already satisfied: idna<4,>=2.5 in /usr/local/lib/python3.10/dist-packages (from requests<3.0.0,>=2.13.0->spacy<3.8.0,>=3.7.2->en-core-web-sm==3.7.1) (3.6)\n",
            "Requirement already satisfied: urllib3<3,>=1.21.1 in /usr/local/lib/python3.10/dist-packages (from requests<3.0.0,>=2.13.0->spacy<3.8.0,>=3.7.2->en-core-web-sm==3.7.1) (2.0.7)\n",
            "Requirement already satisfied: certifi>=2017.4.17 in /usr/local/lib/python3.10/dist-packages (from requests<3.0.0,>=2.13.0->spacy<3.8.0,>=3.7.2->en-core-web-sm==3.7.1) (2024.2.2)\n",
            "Requirement already satisfied: blis<0.8.0,>=0.7.8 in /usr/local/lib/python3.10/dist-packages (from thinc<8.3.0,>=8.1.8->spacy<3.8.0,>=3.7.2->en-core-web-sm==3.7.1) (0.7.11)\n",
            "Requirement already satisfied: confection<1.0.0,>=0.0.1 in /usr/local/lib/python3.10/dist-packages (from thinc<8.3.0,>=8.1.8->spacy<3.8.0,>=3.7.2->en-core-web-sm==3.7.1) (0.1.4)\n",
            "Requirement already satisfied: click<9.0.0,>=7.1.1 in /usr/local/lib/python3.10/dist-packages (from typer<0.10.0,>=0.3.0->spacy<3.8.0,>=3.7.2->en-core-web-sm==3.7.1) (8.1.7)\n",
            "Requirement already satisfied: cloudpathlib<0.17.0,>=0.7.0 in /usr/local/lib/python3.10/dist-packages (from weasel<0.4.0,>=0.1.0->spacy<3.8.0,>=3.7.2->en-core-web-sm==3.7.1) (0.16.0)\n",
            "Requirement already satisfied: MarkupSafe>=2.0 in /usr/local/lib/python3.10/dist-packages (from jinja2->spacy<3.8.0,>=3.7.2->en-core-web-sm==3.7.1) (2.1.5)\n",
            "\u001b[38;5;2m✔ Download and installation successful\u001b[0m\n",
            "You can now load the package via spacy.load('en_core_web_sm')\n",
            "Collecting top2vec[sentence_transformers]\n",
            "  Downloading top2vec-1.0.34-py3-none-any.whl (27 kB)\n",
            "Requirement already satisfied: numpy>=1.20.0 in /usr/local/lib/python3.10/dist-packages (from top2vec[sentence_transformers]) (1.25.2)\n",
            "Requirement already satisfied: pandas in /usr/local/lib/python3.10/dist-packages (from top2vec[sentence_transformers]) (1.5.3)\n",
            "Requirement already satisfied: scikit-learn>=1.2.0 in /usr/local/lib/python3.10/dist-packages (from top2vec[sentence_transformers]) (1.2.2)\n",
            "Requirement already satisfied: gensim>=4.0.0 in /usr/local/lib/python3.10/dist-packages (from top2vec[sentence_transformers]) (4.3.2)\n",
            "Collecting umap-learn>=0.5.1 (from top2vec[sentence_transformers])\n",
            "  Downloading umap-learn-0.5.5.tar.gz (90 kB)\n",
            "\u001b[2K     \u001b[90m━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━\u001b[0m \u001b[32m90.9/90.9 kB\u001b[0m \u001b[31m1.4 MB/s\u001b[0m eta \u001b[36m0:00:00\u001b[0m\n",
            "\u001b[?25h  Preparing metadata (setup.py) ... \u001b[?25l\u001b[?25hdone\n",
            "Collecting hdbscan>=0.8.27 (from top2vec[sentence_transformers])\n",
            "  Downloading hdbscan-0.8.33.tar.gz (5.2 MB)\n",
            "\u001b[2K     \u001b[90m━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━\u001b[0m \u001b[32m5.2/5.2 MB\u001b[0m \u001b[31m20.3 MB/s\u001b[0m eta \u001b[36m0:00:00\u001b[0m\n",
            "\u001b[?25h  Installing build dependencies ... \u001b[?25l\u001b[?25hdone\n",
            "  Getting requirements to build wheel ... \u001b[?25l\u001b[?25hdone\n",
            "  Preparing metadata (pyproject.toml) ... \u001b[?25l\u001b[?25hdone\n",
            "Requirement already satisfied: wordcloud in /usr/local/lib/python3.10/dist-packages (from top2vec[sentence_transformers]) (1.9.3)\n",
            "Requirement already satisfied: torch in /usr/local/lib/python3.10/dist-packages (from top2vec[sentence_transformers]) (2.1.0+cu121)\n",
            "Collecting sentence-transformers (from top2vec[sentence_transformers])\n",
            "  Downloading sentence_transformers-2.3.1-py3-none-any.whl (132 kB)\n",
            "\u001b[2K     \u001b[90m━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━\u001b[0m \u001b[32m132.8/132.8 kB\u001b[0m \u001b[31m12.8 MB/s\u001b[0m eta \u001b[36m0:00:00\u001b[0m\n",
            "\u001b[?25hRequirement already satisfied: scipy>=1.7.0 in /usr/local/lib/python3.10/dist-packages (from gensim>=4.0.0->top2vec[sentence_transformers]) (1.11.4)\n",
            "Requirement already satisfied: smart-open>=1.8.1 in /usr/local/lib/python3.10/dist-packages (from gensim>=4.0.0->top2vec[sentence_transformers]) (6.4.0)\n",
            "Collecting cython<3,>=0.27 (from hdbscan>=0.8.27->top2vec[sentence_transformers])\n",
            "  Using cached Cython-0.29.37-cp310-cp310-manylinux_2_17_x86_64.manylinux2014_x86_64.manylinux_2_24_x86_64.whl (1.9 MB)\n",
            "Requirement already satisfied: joblib>=1.0 in /usr/local/lib/python3.10/dist-packages (from hdbscan>=0.8.27->top2vec[sentence_transformers]) (1.3.2)\n",
            "Requirement already satisfied: threadpoolctl>=2.0.0 in /usr/local/lib/python3.10/dist-packages (from scikit-learn>=1.2.0->top2vec[sentence_transformers]) (3.2.0)\n",
            "Requirement already satisfied: numba>=0.51.2 in /usr/local/lib/python3.10/dist-packages (from umap-learn>=0.5.1->top2vec[sentence_transformers]) (0.58.1)\n",
            "Collecting pynndescent>=0.5 (from umap-learn>=0.5.1->top2vec[sentence_transformers])\n",
            "  Downloading pynndescent-0.5.11-py3-none-any.whl (55 kB)\n",
            "\u001b[2K     \u001b[90m━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━\u001b[0m \u001b[32m55.8/55.8 kB\u001b[0m \u001b[31m6.3 MB/s\u001b[0m eta \u001b[36m0:00:00\u001b[0m\n",
            "\u001b[?25hRequirement already satisfied: tqdm in /usr/local/lib/python3.10/dist-packages (from umap-learn>=0.5.1->top2vec[sentence_transformers]) (4.66.2)\n",
            "Requirement already satisfied: python-dateutil>=2.8.1 in /usr/local/lib/python3.10/dist-packages (from pandas->top2vec[sentence_transformers]) (2.8.2)\n",
            "Requirement already satisfied: pytz>=2020.1 in /usr/local/lib/python3.10/dist-packages (from pandas->top2vec[sentence_transformers]) (2023.4)\n",
            "Requirement already satisfied: transformers<5.0.0,>=4.32.0 in /usr/local/lib/python3.10/dist-packages (from sentence-transformers->top2vec[sentence_transformers]) (4.35.2)\n",
            "Requirement already satisfied: nltk in /usr/local/lib/python3.10/dist-packages (from sentence-transformers->top2vec[sentence_transformers]) (3.8.1)\n",
            "Requirement already satisfied: sentencepiece in /usr/local/lib/python3.10/dist-packages (from sentence-transformers->top2vec[sentence_transformers]) (0.1.99)\n",
            "Requirement already satisfied: huggingface-hub>=0.15.1 in /usr/local/lib/python3.10/dist-packages (from sentence-transformers->top2vec[sentence_transformers]) (0.20.3)\n",
            "Requirement already satisfied: Pillow in /usr/local/lib/python3.10/dist-packages (from sentence-transformers->top2vec[sentence_transformers]) (9.4.0)\n",
            "Requirement already satisfied: filelock in /usr/local/lib/python3.10/dist-packages (from torch->top2vec[sentence_transformers]) (3.13.1)\n",
            "Requirement already satisfied: typing-extensions in /usr/local/lib/python3.10/dist-packages (from torch->top2vec[sentence_transformers]) (4.9.0)\n",
            "Requirement already satisfied: sympy in /usr/local/lib/python3.10/dist-packages (from torch->top2vec[sentence_transformers]) (1.12)\n",
            "Requirement already satisfied: networkx in /usr/local/lib/python3.10/dist-packages (from torch->top2vec[sentence_transformers]) (3.2.1)\n",
            "Requirement already satisfied: jinja2 in /usr/local/lib/python3.10/dist-packages (from torch->top2vec[sentence_transformers]) (3.1.3)\n",
            "Requirement already satisfied: fsspec in /usr/local/lib/python3.10/dist-packages (from torch->top2vec[sentence_transformers]) (2023.6.0)\n",
            "Requirement already satisfied: triton==2.1.0 in /usr/local/lib/python3.10/dist-packages (from torch->top2vec[sentence_transformers]) (2.1.0)\n",
            "Requirement already satisfied: matplotlib in /usr/local/lib/python3.10/dist-packages (from wordcloud->top2vec[sentence_transformers]) (3.7.1)\n",
            "Requirement already satisfied: requests in /usr/local/lib/python3.10/dist-packages (from huggingface-hub>=0.15.1->sentence-transformers->top2vec[sentence_transformers]) (2.31.0)\n",
            "Requirement already satisfied: pyyaml>=5.1 in /usr/local/lib/python3.10/dist-packages (from huggingface-hub>=0.15.1->sentence-transformers->top2vec[sentence_transformers]) (6.0.1)\n",
            "Requirement already satisfied: packaging>=20.9 in /usr/local/lib/python3.10/dist-packages (from huggingface-hub>=0.15.1->sentence-transformers->top2vec[sentence_transformers]) (23.2)\n",
            "Requirement already satisfied: llvmlite<0.42,>=0.41.0dev0 in /usr/local/lib/python3.10/dist-packages (from numba>=0.51.2->umap-learn>=0.5.1->top2vec[sentence_transformers]) (0.41.1)\n",
            "Requirement already satisfied: six>=1.5 in /usr/local/lib/python3.10/dist-packages (from python-dateutil>=2.8.1->pandas->top2vec[sentence_transformers]) (1.16.0)\n",
            "Requirement already satisfied: regex!=2019.12.17 in /usr/local/lib/python3.10/dist-packages (from transformers<5.0.0,>=4.32.0->sentence-transformers->top2vec[sentence_transformers]) (2023.12.25)\n",
            "Requirement already satisfied: tokenizers<0.19,>=0.14 in /usr/local/lib/python3.10/dist-packages (from transformers<5.0.0,>=4.32.0->sentence-transformers->top2vec[sentence_transformers]) (0.15.2)\n",
            "Requirement already satisfied: safetensors>=0.3.1 in /usr/local/lib/python3.10/dist-packages (from transformers<5.0.0,>=4.32.0->sentence-transformers->top2vec[sentence_transformers]) (0.4.2)\n",
            "Requirement already satisfied: MarkupSafe>=2.0 in /usr/local/lib/python3.10/dist-packages (from jinja2->torch->top2vec[sentence_transformers]) (2.1.5)\n",
            "Requirement already satisfied: contourpy>=1.0.1 in /usr/local/lib/python3.10/dist-packages (from matplotlib->wordcloud->top2vec[sentence_transformers]) (1.2.0)\n",
            "Requirement already satisfied: cycler>=0.10 in /usr/local/lib/python3.10/dist-packages (from matplotlib->wordcloud->top2vec[sentence_transformers]) (0.12.1)\n",
            "Requirement already satisfied: fonttools>=4.22.0 in /usr/local/lib/python3.10/dist-packages (from matplotlib->wordcloud->top2vec[sentence_transformers]) (4.48.1)\n",
            "Requirement already satisfied: kiwisolver>=1.0.1 in /usr/local/lib/python3.10/dist-packages (from matplotlib->wordcloud->top2vec[sentence_transformers]) (1.4.5)\n",
            "Requirement already satisfied: pyparsing>=2.3.1 in /usr/local/lib/python3.10/dist-packages (from matplotlib->wordcloud->top2vec[sentence_transformers]) (3.1.1)\n",
            "Requirement already satisfied: click in /usr/local/lib/python3.10/dist-packages (from nltk->sentence-transformers->top2vec[sentence_transformers]) (8.1.7)\n",
            "Requirement already satisfied: mpmath>=0.19 in /usr/local/lib/python3.10/dist-packages (from sympy->torch->top2vec[sentence_transformers]) (1.3.0)\n",
            "Requirement already satisfied: charset-normalizer<4,>=2 in /usr/local/lib/python3.10/dist-packages (from requests->huggingface-hub>=0.15.1->sentence-transformers->top2vec[sentence_transformers]) (3.3.2)\n",
            "Requirement already satisfied: idna<4,>=2.5 in /usr/local/lib/python3.10/dist-packages (from requests->huggingface-hub>=0.15.1->sentence-transformers->top2vec[sentence_transformers]) (3.6)\n",
            "Requirement already satisfied: urllib3<3,>=1.21.1 in /usr/local/lib/python3.10/dist-packages (from requests->huggingface-hub>=0.15.1->sentence-transformers->top2vec[sentence_transformers]) (2.0.7)\n",
            "Requirement already satisfied: certifi>=2017.4.17 in /usr/local/lib/python3.10/dist-packages (from requests->huggingface-hub>=0.15.1->sentence-transformers->top2vec[sentence_transformers]) (2024.2.2)\n",
            "Building wheels for collected packages: hdbscan, umap-learn\n",
            "  Building wheel for hdbscan (pyproject.toml) ... \u001b[?25l\u001b[?25hdone\n",
            "  Created wheel for hdbscan: filename=hdbscan-0.8.33-cp310-cp310-linux_x86_64.whl size=3039299 sha256=c6b8951db45f4477b1b17c057fcce006b6a2156f9d37031a329951a720179218\n",
            "  Stored in directory: /root/.cache/pip/wheels/75/0b/3b/dc4f60b7cc455efaefb62883a7483e76f09d06ca81cf87d610\n",
            "  Building wheel for umap-learn (setup.py) ... \u001b[?25l\u001b[?25hdone\n",
            "  Created wheel for umap-learn: filename=umap_learn-0.5.5-py3-none-any.whl size=86832 sha256=b92ccbe102fc5d51bd4f1d5c2740329302703e2c1e39932226c46f8aed551fb4\n",
            "  Stored in directory: /root/.cache/pip/wheels/3a/70/07/428d2b58660a1a3b431db59b806a10da736612ebbc66c1bcc5\n",
            "Successfully built hdbscan umap-learn\n",
            "Installing collected packages: cython, pynndescent, hdbscan, umap-learn, top2vec, sentence-transformers\n",
            "  Attempting uninstall: cython\n",
            "    Found existing installation: Cython 3.0.8\n",
            "    Uninstalling Cython-3.0.8:\n",
            "      Successfully uninstalled Cython-3.0.8\n",
            "Successfully installed cython-0.29.37 hdbscan-0.8.33 pynndescent-0.5.11 sentence-transformers-2.3.1 top2vec-1.0.34 umap-learn-0.5.5\n",
            "Requirement already satisfied: top2vec[sentence_encoders] in /usr/local/lib/python3.10/dist-packages (1.0.34)\n",
            "Requirement already satisfied: numpy>=1.20.0 in /usr/local/lib/python3.10/dist-packages (from top2vec[sentence_encoders]) (1.25.2)\n",
            "Requirement already satisfied: pandas in /usr/local/lib/python3.10/dist-packages (from top2vec[sentence_encoders]) (1.5.3)\n",
            "Requirement already satisfied: scikit-learn>=1.2.0 in /usr/local/lib/python3.10/dist-packages (from top2vec[sentence_encoders]) (1.2.2)\n",
            "Requirement already satisfied: gensim>=4.0.0 in /usr/local/lib/python3.10/dist-packages (from top2vec[sentence_encoders]) (4.3.2)\n",
            "Requirement already satisfied: umap-learn>=0.5.1 in /usr/local/lib/python3.10/dist-packages (from top2vec[sentence_encoders]) (0.5.5)\n",
            "Requirement already satisfied: hdbscan>=0.8.27 in /usr/local/lib/python3.10/dist-packages (from top2vec[sentence_encoders]) (0.8.33)\n",
            "Requirement already satisfied: wordcloud in /usr/local/lib/python3.10/dist-packages (from top2vec[sentence_encoders]) (1.9.3)\n",
            "Requirement already satisfied: tensorflow in /usr/local/lib/python3.10/dist-packages (from top2vec[sentence_encoders]) (2.15.0)\n",
            "Requirement already satisfied: tensorflow-hub in /usr/local/lib/python3.10/dist-packages (from top2vec[sentence_encoders]) (0.16.1)\n",
            "Collecting tensorflow-text (from top2vec[sentence_encoders])\n",
            "  Downloading tensorflow_text-2.15.0-cp310-cp310-manylinux_2_17_x86_64.manylinux2014_x86_64.whl (5.2 MB)\n",
            "\u001b[2K     \u001b[90m━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━\u001b[0m \u001b[32m5.2/5.2 MB\u001b[0m \u001b[31m17.2 MB/s\u001b[0m eta \u001b[36m0:00:00\u001b[0m\n",
            "\u001b[?25hRequirement already satisfied: scipy>=1.7.0 in /usr/local/lib/python3.10/dist-packages (from gensim>=4.0.0->top2vec[sentence_encoders]) (1.11.4)\n",
            "Requirement already satisfied: smart-open>=1.8.1 in /usr/local/lib/python3.10/dist-packages (from gensim>=4.0.0->top2vec[sentence_encoders]) (6.4.0)\n",
            "Requirement already satisfied: cython<3,>=0.27 in /usr/local/lib/python3.10/dist-packages (from hdbscan>=0.8.27->top2vec[sentence_encoders]) (0.29.37)\n",
            "Requirement already satisfied: joblib>=1.0 in /usr/local/lib/python3.10/dist-packages (from hdbscan>=0.8.27->top2vec[sentence_encoders]) (1.3.2)\n",
            "Requirement already satisfied: threadpoolctl>=2.0.0 in /usr/local/lib/python3.10/dist-packages (from scikit-learn>=1.2.0->top2vec[sentence_encoders]) (3.2.0)\n",
            "Requirement already satisfied: numba>=0.51.2 in /usr/local/lib/python3.10/dist-packages (from umap-learn>=0.5.1->top2vec[sentence_encoders]) (0.58.1)\n",
            "Requirement already satisfied: pynndescent>=0.5 in /usr/local/lib/python3.10/dist-packages (from umap-learn>=0.5.1->top2vec[sentence_encoders]) (0.5.11)\n",
            "Requirement already satisfied: tqdm in /usr/local/lib/python3.10/dist-packages (from umap-learn>=0.5.1->top2vec[sentence_encoders]) (4.66.2)\n",
            "Requirement already satisfied: python-dateutil>=2.8.1 in /usr/local/lib/python3.10/dist-packages (from pandas->top2vec[sentence_encoders]) (2.8.2)\n",
            "Requirement already satisfied: pytz>=2020.1 in /usr/local/lib/python3.10/dist-packages (from pandas->top2vec[sentence_encoders]) (2023.4)\n",
            "Requirement already satisfied: absl-py>=1.0.0 in /usr/local/lib/python3.10/dist-packages (from tensorflow->top2vec[sentence_encoders]) (1.4.0)\n",
            "Requirement already satisfied: astunparse>=1.6.0 in /usr/local/lib/python3.10/dist-packages (from tensorflow->top2vec[sentence_encoders]) (1.6.3)\n",
            "Requirement already satisfied: flatbuffers>=23.5.26 in /usr/local/lib/python3.10/dist-packages (from tensorflow->top2vec[sentence_encoders]) (23.5.26)\n",
            "Requirement already satisfied: gast!=0.5.0,!=0.5.1,!=0.5.2,>=0.2.1 in /usr/local/lib/python3.10/dist-packages (from tensorflow->top2vec[sentence_encoders]) (0.5.4)\n",
            "Requirement already satisfied: google-pasta>=0.1.1 in /usr/local/lib/python3.10/dist-packages (from tensorflow->top2vec[sentence_encoders]) (0.2.0)\n",
            "Requirement already satisfied: h5py>=2.9.0 in /usr/local/lib/python3.10/dist-packages (from tensorflow->top2vec[sentence_encoders]) (3.9.0)\n",
            "Requirement already satisfied: libclang>=13.0.0 in /usr/local/lib/python3.10/dist-packages (from tensorflow->top2vec[sentence_encoders]) (16.0.6)\n",
            "Requirement already satisfied: ml-dtypes~=0.2.0 in /usr/local/lib/python3.10/dist-packages (from tensorflow->top2vec[sentence_encoders]) (0.2.0)\n",
            "Requirement already satisfied: opt-einsum>=2.3.2 in /usr/local/lib/python3.10/dist-packages (from tensorflow->top2vec[sentence_encoders]) (3.3.0)\n",
            "Requirement already satisfied: packaging in /usr/local/lib/python3.10/dist-packages (from tensorflow->top2vec[sentence_encoders]) (23.2)\n",
            "Requirement already satisfied: protobuf!=4.21.0,!=4.21.1,!=4.21.2,!=4.21.3,!=4.21.4,!=4.21.5,<5.0.0dev,>=3.20.3 in /usr/local/lib/python3.10/dist-packages (from tensorflow->top2vec[sentence_encoders]) (3.20.3)\n",
            "Requirement already satisfied: setuptools in /usr/local/lib/python3.10/dist-packages (from tensorflow->top2vec[sentence_encoders]) (67.7.2)\n",
            "Requirement already satisfied: six>=1.12.0 in /usr/local/lib/python3.10/dist-packages (from tensorflow->top2vec[sentence_encoders]) (1.16.0)\n",
            "Requirement already satisfied: termcolor>=1.1.0 in /usr/local/lib/python3.10/dist-packages (from tensorflow->top2vec[sentence_encoders]) (2.4.0)\n",
            "Requirement already satisfied: typing-extensions>=3.6.6 in /usr/local/lib/python3.10/dist-packages (from tensorflow->top2vec[sentence_encoders]) (4.9.0)\n",
            "Requirement already satisfied: wrapt<1.15,>=1.11.0 in /usr/local/lib/python3.10/dist-packages (from tensorflow->top2vec[sentence_encoders]) (1.14.1)\n",
            "Requirement already satisfied: tensorflow-io-gcs-filesystem>=0.23.1 in /usr/local/lib/python3.10/dist-packages (from tensorflow->top2vec[sentence_encoders]) (0.36.0)\n",
            "Requirement already satisfied: grpcio<2.0,>=1.24.3 in /usr/local/lib/python3.10/dist-packages (from tensorflow->top2vec[sentence_encoders]) (1.60.1)\n",
            "Requirement already satisfied: tensorboard<2.16,>=2.15 in /usr/local/lib/python3.10/dist-packages (from tensorflow->top2vec[sentence_encoders]) (2.15.2)\n",
            "Requirement already satisfied: tensorflow-estimator<2.16,>=2.15.0 in /usr/local/lib/python3.10/dist-packages (from tensorflow->top2vec[sentence_encoders]) (2.15.0)\n",
            "Requirement already satisfied: keras<2.16,>=2.15.0 in /usr/local/lib/python3.10/dist-packages (from tensorflow->top2vec[sentence_encoders]) (2.15.0)\n",
            "Requirement already satisfied: tf-keras>=2.14.1 in /usr/local/lib/python3.10/dist-packages (from tensorflow-hub->top2vec[sentence_encoders]) (2.15.0)\n",
            "Requirement already satisfied: pillow in /usr/local/lib/python3.10/dist-packages (from wordcloud->top2vec[sentence_encoders]) (9.4.0)\n",
            "Requirement already satisfied: matplotlib in /usr/local/lib/python3.10/dist-packages (from wordcloud->top2vec[sentence_encoders]) (3.7.1)\n",
            "Requirement already satisfied: wheel<1.0,>=0.23.0 in /usr/local/lib/python3.10/dist-packages (from astunparse>=1.6.0->tensorflow->top2vec[sentence_encoders]) (0.42.0)\n",
            "Requirement already satisfied: llvmlite<0.42,>=0.41.0dev0 in /usr/local/lib/python3.10/dist-packages (from numba>=0.51.2->umap-learn>=0.5.1->top2vec[sentence_encoders]) (0.41.1)\n",
            "Requirement already satisfied: google-auth<3,>=1.6.3 in /usr/local/lib/python3.10/dist-packages (from tensorboard<2.16,>=2.15->tensorflow->top2vec[sentence_encoders]) (2.27.0)\n",
            "Requirement already satisfied: google-auth-oauthlib<2,>=0.5 in /usr/local/lib/python3.10/dist-packages (from tensorboard<2.16,>=2.15->tensorflow->top2vec[sentence_encoders]) (1.2.0)\n",
            "Requirement already satisfied: markdown>=2.6.8 in /usr/local/lib/python3.10/dist-packages (from tensorboard<2.16,>=2.15->tensorflow->top2vec[sentence_encoders]) (3.5.2)\n",
            "Requirement already satisfied: requests<3,>=2.21.0 in /usr/local/lib/python3.10/dist-packages (from tensorboard<2.16,>=2.15->tensorflow->top2vec[sentence_encoders]) (2.31.0)\n",
            "Requirement already satisfied: tensorboard-data-server<0.8.0,>=0.7.0 in /usr/local/lib/python3.10/dist-packages (from tensorboard<2.16,>=2.15->tensorflow->top2vec[sentence_encoders]) (0.7.2)\n",
            "Requirement already satisfied: werkzeug>=1.0.1 in /usr/local/lib/python3.10/dist-packages (from tensorboard<2.16,>=2.15->tensorflow->top2vec[sentence_encoders]) (3.0.1)\n",
            "Requirement already satisfied: contourpy>=1.0.1 in /usr/local/lib/python3.10/dist-packages (from matplotlib->wordcloud->top2vec[sentence_encoders]) (1.2.0)\n",
            "Requirement already satisfied: cycler>=0.10 in /usr/local/lib/python3.10/dist-packages (from matplotlib->wordcloud->top2vec[sentence_encoders]) (0.12.1)\n",
            "Requirement already satisfied: fonttools>=4.22.0 in /usr/local/lib/python3.10/dist-packages (from matplotlib->wordcloud->top2vec[sentence_encoders]) (4.48.1)\n",
            "Requirement already satisfied: kiwisolver>=1.0.1 in /usr/local/lib/python3.10/dist-packages (from matplotlib->wordcloud->top2vec[sentence_encoders]) (1.4.5)\n",
            "Requirement already satisfied: pyparsing>=2.3.1 in /usr/local/lib/python3.10/dist-packages (from matplotlib->wordcloud->top2vec[sentence_encoders]) (3.1.1)\n",
            "Requirement already satisfied: cachetools<6.0,>=2.0.0 in /usr/local/lib/python3.10/dist-packages (from google-auth<3,>=1.6.3->tensorboard<2.16,>=2.15->tensorflow->top2vec[sentence_encoders]) (5.3.2)\n",
            "Requirement already satisfied: pyasn1-modules>=0.2.1 in /usr/local/lib/python3.10/dist-packages (from google-auth<3,>=1.6.3->tensorboard<2.16,>=2.15->tensorflow->top2vec[sentence_encoders]) (0.3.0)\n",
            "Requirement already satisfied: rsa<5,>=3.1.4 in /usr/local/lib/python3.10/dist-packages (from google-auth<3,>=1.6.3->tensorboard<2.16,>=2.15->tensorflow->top2vec[sentence_encoders]) (4.9)\n",
            "Requirement already satisfied: requests-oauthlib>=0.7.0 in /usr/local/lib/python3.10/dist-packages (from google-auth-oauthlib<2,>=0.5->tensorboard<2.16,>=2.15->tensorflow->top2vec[sentence_encoders]) (1.3.1)\n",
            "Requirement already satisfied: charset-normalizer<4,>=2 in /usr/local/lib/python3.10/dist-packages (from requests<3,>=2.21.0->tensorboard<2.16,>=2.15->tensorflow->top2vec[sentence_encoders]) (3.3.2)\n",
            "Requirement already satisfied: idna<4,>=2.5 in /usr/local/lib/python3.10/dist-packages (from requests<3,>=2.21.0->tensorboard<2.16,>=2.15->tensorflow->top2vec[sentence_encoders]) (3.6)\n",
            "Requirement already satisfied: urllib3<3,>=1.21.1 in /usr/local/lib/python3.10/dist-packages (from requests<3,>=2.21.0->tensorboard<2.16,>=2.15->tensorflow->top2vec[sentence_encoders]) (2.0.7)\n",
            "Requirement already satisfied: certifi>=2017.4.17 in /usr/local/lib/python3.10/dist-packages (from requests<3,>=2.21.0->tensorboard<2.16,>=2.15->tensorflow->top2vec[sentence_encoders]) (2024.2.2)\n",
            "Requirement already satisfied: MarkupSafe>=2.1.1 in /usr/local/lib/python3.10/dist-packages (from werkzeug>=1.0.1->tensorboard<2.16,>=2.15->tensorflow->top2vec[sentence_encoders]) (2.1.5)\n",
            "Requirement already satisfied: pyasn1<0.6.0,>=0.4.6 in /usr/local/lib/python3.10/dist-packages (from pyasn1-modules>=0.2.1->google-auth<3,>=1.6.3->tensorboard<2.16,>=2.15->tensorflow->top2vec[sentence_encoders]) (0.5.1)\n",
            "Requirement already satisfied: oauthlib>=3.0.0 in /usr/local/lib/python3.10/dist-packages (from requests-oauthlib>=0.7.0->google-auth-oauthlib<2,>=0.5->tensorboard<2.16,>=2.15->tensorflow->top2vec[sentence_encoders]) (3.2.2)\n",
            "Installing collected packages: tensorflow-text\n",
            "Successfully installed tensorflow-text-2.15.0\n",
            "Requirement already satisfied: tensorflow in /usr/local/lib/python3.10/dist-packages (2.15.0)\n",
            "Requirement already satisfied: tensorflow_hub in /usr/local/lib/python3.10/dist-packages (0.16.1)\n",
            "Requirement already satisfied: tensorflow_text in /usr/local/lib/python3.10/dist-packages (2.15.0)\n",
            "Requirement already satisfied: absl-py>=1.0.0 in /usr/local/lib/python3.10/dist-packages (from tensorflow) (1.4.0)\n",
            "Requirement already satisfied: astunparse>=1.6.0 in /usr/local/lib/python3.10/dist-packages (from tensorflow) (1.6.3)\n",
            "Requirement already satisfied: flatbuffers>=23.5.26 in /usr/local/lib/python3.10/dist-packages (from tensorflow) (23.5.26)\n",
            "Requirement already satisfied: gast!=0.5.0,!=0.5.1,!=0.5.2,>=0.2.1 in /usr/local/lib/python3.10/dist-packages (from tensorflow) (0.5.4)\n",
            "Requirement already satisfied: google-pasta>=0.1.1 in /usr/local/lib/python3.10/dist-packages (from tensorflow) (0.2.0)\n",
            "Requirement already satisfied: h5py>=2.9.0 in /usr/local/lib/python3.10/dist-packages (from tensorflow) (3.9.0)\n",
            "Requirement already satisfied: libclang>=13.0.0 in /usr/local/lib/python3.10/dist-packages (from tensorflow) (16.0.6)\n",
            "Requirement already satisfied: ml-dtypes~=0.2.0 in /usr/local/lib/python3.10/dist-packages (from tensorflow) (0.2.0)\n",
            "Requirement already satisfied: numpy<2.0.0,>=1.23.5 in /usr/local/lib/python3.10/dist-packages (from tensorflow) (1.25.2)\n",
            "Requirement already satisfied: opt-einsum>=2.3.2 in /usr/local/lib/python3.10/dist-packages (from tensorflow) (3.3.0)\n",
            "Requirement already satisfied: packaging in /usr/local/lib/python3.10/dist-packages (from tensorflow) (23.2)\n",
            "Requirement already satisfied: protobuf!=4.21.0,!=4.21.1,!=4.21.2,!=4.21.3,!=4.21.4,!=4.21.5,<5.0.0dev,>=3.20.3 in /usr/local/lib/python3.10/dist-packages (from tensorflow) (3.20.3)\n",
            "Requirement already satisfied: setuptools in /usr/local/lib/python3.10/dist-packages (from tensorflow) (67.7.2)\n",
            "Requirement already satisfied: six>=1.12.0 in /usr/local/lib/python3.10/dist-packages (from tensorflow) (1.16.0)\n",
            "Requirement already satisfied: termcolor>=1.1.0 in /usr/local/lib/python3.10/dist-packages (from tensorflow) (2.4.0)\n",
            "Requirement already satisfied: typing-extensions>=3.6.6 in /usr/local/lib/python3.10/dist-packages (from tensorflow) (4.9.0)\n",
            "Requirement already satisfied: wrapt<1.15,>=1.11.0 in /usr/local/lib/python3.10/dist-packages (from tensorflow) (1.14.1)\n",
            "Requirement already satisfied: tensorflow-io-gcs-filesystem>=0.23.1 in /usr/local/lib/python3.10/dist-packages (from tensorflow) (0.36.0)\n",
            "Requirement already satisfied: grpcio<2.0,>=1.24.3 in /usr/local/lib/python3.10/dist-packages (from tensorflow) (1.60.1)\n",
            "Requirement already satisfied: tensorboard<2.16,>=2.15 in /usr/local/lib/python3.10/dist-packages (from tensorflow) (2.15.2)\n",
            "Requirement already satisfied: tensorflow-estimator<2.16,>=2.15.0 in /usr/local/lib/python3.10/dist-packages (from tensorflow) (2.15.0)\n",
            "Requirement already satisfied: keras<2.16,>=2.15.0 in /usr/local/lib/python3.10/dist-packages (from tensorflow) (2.15.0)\n",
            "Requirement already satisfied: tf-keras>=2.14.1 in /usr/local/lib/python3.10/dist-packages (from tensorflow_hub) (2.15.0)\n",
            "Requirement already satisfied: wheel<1.0,>=0.23.0 in /usr/local/lib/python3.10/dist-packages (from astunparse>=1.6.0->tensorflow) (0.42.0)\n",
            "Requirement already satisfied: google-auth<3,>=1.6.3 in /usr/local/lib/python3.10/dist-packages (from tensorboard<2.16,>=2.15->tensorflow) (2.27.0)\n",
            "Requirement already satisfied: google-auth-oauthlib<2,>=0.5 in /usr/local/lib/python3.10/dist-packages (from tensorboard<2.16,>=2.15->tensorflow) (1.2.0)\n",
            "Requirement already satisfied: markdown>=2.6.8 in /usr/local/lib/python3.10/dist-packages (from tensorboard<2.16,>=2.15->tensorflow) (3.5.2)\n",
            "Requirement already satisfied: requests<3,>=2.21.0 in /usr/local/lib/python3.10/dist-packages (from tensorboard<2.16,>=2.15->tensorflow) (2.31.0)\n",
            "Requirement already satisfied: tensorboard-data-server<0.8.0,>=0.7.0 in /usr/local/lib/python3.10/dist-packages (from tensorboard<2.16,>=2.15->tensorflow) (0.7.2)\n",
            "Requirement already satisfied: werkzeug>=1.0.1 in /usr/local/lib/python3.10/dist-packages (from tensorboard<2.16,>=2.15->tensorflow) (3.0.1)\n",
            "Requirement already satisfied: cachetools<6.0,>=2.0.0 in /usr/local/lib/python3.10/dist-packages (from google-auth<3,>=1.6.3->tensorboard<2.16,>=2.15->tensorflow) (5.3.2)\n",
            "Requirement already satisfied: pyasn1-modules>=0.2.1 in /usr/local/lib/python3.10/dist-packages (from google-auth<3,>=1.6.3->tensorboard<2.16,>=2.15->tensorflow) (0.3.0)\n",
            "Requirement already satisfied: rsa<5,>=3.1.4 in /usr/local/lib/python3.10/dist-packages (from google-auth<3,>=1.6.3->tensorboard<2.16,>=2.15->tensorflow) (4.9)\n",
            "Requirement already satisfied: requests-oauthlib>=0.7.0 in /usr/local/lib/python3.10/dist-packages (from google-auth-oauthlib<2,>=0.5->tensorboard<2.16,>=2.15->tensorflow) (1.3.1)\n",
            "Requirement already satisfied: charset-normalizer<4,>=2 in /usr/local/lib/python3.10/dist-packages (from requests<3,>=2.21.0->tensorboard<2.16,>=2.15->tensorflow) (3.3.2)\n",
            "Requirement already satisfied: idna<4,>=2.5 in /usr/local/lib/python3.10/dist-packages (from requests<3,>=2.21.0->tensorboard<2.16,>=2.15->tensorflow) (3.6)\n",
            "Requirement already satisfied: urllib3<3,>=1.21.1 in /usr/local/lib/python3.10/dist-packages (from requests<3,>=2.21.0->tensorboard<2.16,>=2.15->tensorflow) (2.0.7)\n",
            "Requirement already satisfied: certifi>=2017.4.17 in /usr/local/lib/python3.10/dist-packages (from requests<3,>=2.21.0->tensorboard<2.16,>=2.15->tensorflow) (2024.2.2)\n",
            "Requirement already satisfied: MarkupSafe>=2.1.1 in /usr/local/lib/python3.10/dist-packages (from werkzeug>=1.0.1->tensorboard<2.16,>=2.15->tensorflow) (2.1.5)\n",
            "Requirement already satisfied: pyasn1<0.6.0,>=0.4.6 in /usr/local/lib/python3.10/dist-packages (from pyasn1-modules>=0.2.1->google-auth<3,>=1.6.3->tensorboard<2.16,>=2.15->tensorflow) (0.5.1)\n",
            "Requirement already satisfied: oauthlib>=3.0.0 in /usr/local/lib/python3.10/dist-packages (from requests-oauthlib>=0.7.0->google-auth-oauthlib<2,>=0.5->tensorboard<2.16,>=2.15->tensorflow) (3.2.2)\n",
            "Collecting colored\n",
            "  Downloading colored-2.2.4-py3-none-any.whl (16 kB)\n",
            "Installing collected packages: colored\n",
            "Successfully installed colored-2.2.4\n"
          ]
        }
      ]
    },
    {
      "cell_type": "code",
      "source": [
        "import pandas as pd\n",
        "import spacy\n",
        "from spacy.matcher import Matcher\n",
        "nlp = spacy.load(\"en_core_web_sm\")\n",
        "import numpy as np\n",
        "from top2vec import Top2Vec\n",
        "from colored import Fore, Back, Style"
      ],
      "metadata": {
        "id": "PzEAqoOOAmMD"
      },
      "execution_count": 2,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [
        "df = pd.read_csv('/content/biology_new.csv')\n",
        "df = df[['text', 'title', 'date', 'link', 'cleaned_text', 'fully_cleaned_text']]\n",
        "df.head()"
      ],
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/",
          "height": 293
        },
        "id": "zdbmBtITZvX8",
        "outputId": "2b142a54-d6ee-4bfa-cbc3-0baa8f54c989"
      },
      "execution_count": 3,
      "outputs": [
        {
          "output_type": "execute_result",
          "data": {
            "text/plain": [
              "                                                text  \\\n",
              "0  FOR THE past four billion years or so the only...   \n",
              "1  IN A former leatherworks just off Euston Road ...   \n",
              "2  “How many cells are there in a human being?” I...   \n",
              "3  LIVING creatures are jolly useful. Farmers rea...   \n",
              "4  A broken brain, hidden inside a skull, is hard...   \n",
              "\n",
              "                                               title            date  \\\n",
              "0        The promise and perils of synthetic biology    Apr 4th 2019   \n",
              "1  Will artificial intelligence help to crack bio...    Jan 7th 2017   \n",
              "2  The idea of “holobionts” represents a paradigm...   Jun 14th 2023   \n",
              "3        The remarkable promise of cell-free biology    May 4th 2017   \n",
              "4  Better brain biology will deliver better medic...   Sep 21st 2022   \n",
              "\n",
              "                                                link  \\\n",
              "0  https://www.economist.com/leaders/2019/04/04/t...   \n",
              "1  https://www.economist.com/science-and-technolo...   \n",
              "2  https://www.economist.com/science-and-technolo...   \n",
              "3  https://www.economist.com/leaders/2017/05/04/t...   \n",
              "4  https://www.economist.com/technology-quarterly...   \n",
              "\n",
              "                                        cleaned_text  \\\n",
              "0  for the past four billion years or so the only...   \n",
              "1  in a former leatherworks just off euston road ...   \n",
              "2   how many cells are there in a human being  it...   \n",
              "3  living creatures are jolly useful farmers rear...   \n",
              "4  a broken brain hidden inside a skull is harder...   \n",
              "\n",
              "                                  fully_cleaned_text  \n",
              "0  past four billion years way life earth produce...  \n",
              "1  former leatherworks euston road london hopeful...  \n",
              "2  many cells human sounds like question nerdy pu...  \n",
              "3  living creatures jolly useful farmers rear ani...  \n",
              "4  broken brain hidden inside skull harder diagno...  "
            ],
            "text/html": [
              "\n",
              "  <div id=\"df-61a1b2a3-5d60-482b-8fcd-942aee971b4f\" class=\"colab-df-container\">\n",
              "    <div>\n",
              "<style scoped>\n",
              "    .dataframe tbody tr th:only-of-type {\n",
              "        vertical-align: middle;\n",
              "    }\n",
              "\n",
              "    .dataframe tbody tr th {\n",
              "        vertical-align: top;\n",
              "    }\n",
              "\n",
              "    .dataframe thead th {\n",
              "        text-align: right;\n",
              "    }\n",
              "</style>\n",
              "<table border=\"1\" class=\"dataframe\">\n",
              "  <thead>\n",
              "    <tr style=\"text-align: right;\">\n",
              "      <th></th>\n",
              "      <th>text</th>\n",
              "      <th>title</th>\n",
              "      <th>date</th>\n",
              "      <th>link</th>\n",
              "      <th>cleaned_text</th>\n",
              "      <th>fully_cleaned_text</th>\n",
              "    </tr>\n",
              "  </thead>\n",
              "  <tbody>\n",
              "    <tr>\n",
              "      <th>0</th>\n",
              "      <td>FOR THE past four billion years or so the only...</td>\n",
              "      <td>The promise and perils of synthetic biology</td>\n",
              "      <td>Apr 4th 2019</td>\n",
              "      <td>https://www.economist.com/leaders/2019/04/04/t...</td>\n",
              "      <td>for the past four billion years or so the only...</td>\n",
              "      <td>past four billion years way life earth produce...</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>1</th>\n",
              "      <td>IN A former leatherworks just off Euston Road ...</td>\n",
              "      <td>Will artificial intelligence help to crack bio...</td>\n",
              "      <td>Jan 7th 2017</td>\n",
              "      <td>https://www.economist.com/science-and-technolo...</td>\n",
              "      <td>in a former leatherworks just off euston road ...</td>\n",
              "      <td>former leatherworks euston road london hopeful...</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>2</th>\n",
              "      <td>“How many cells are there in a human being?” I...</td>\n",
              "      <td>The idea of “holobionts” represents a paradigm...</td>\n",
              "      <td>Jun 14th 2023</td>\n",
              "      <td>https://www.economist.com/science-and-technolo...</td>\n",
              "      <td>how many cells are there in a human being  it...</td>\n",
              "      <td>many cells human sounds like question nerdy pu...</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>3</th>\n",
              "      <td>LIVING creatures are jolly useful. Farmers rea...</td>\n",
              "      <td>The remarkable promise of cell-free biology</td>\n",
              "      <td>May 4th 2017</td>\n",
              "      <td>https://www.economist.com/leaders/2017/05/04/t...</td>\n",
              "      <td>living creatures are jolly useful farmers rear...</td>\n",
              "      <td>living creatures jolly useful farmers rear ani...</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>4</th>\n",
              "      <td>A broken brain, hidden inside a skull, is hard...</td>\n",
              "      <td>Better brain biology will deliver better medic...</td>\n",
              "      <td>Sep 21st 2022</td>\n",
              "      <td>https://www.economist.com/technology-quarterly...</td>\n",
              "      <td>a broken brain hidden inside a skull is harder...</td>\n",
              "      <td>broken brain hidden inside skull harder diagno...</td>\n",
              "    </tr>\n",
              "  </tbody>\n",
              "</table>\n",
              "</div>\n",
              "    <div class=\"colab-df-buttons\">\n",
              "\n",
              "  <div class=\"colab-df-container\">\n",
              "    <button class=\"colab-df-convert\" onclick=\"convertToInteractive('df-61a1b2a3-5d60-482b-8fcd-942aee971b4f')\"\n",
              "            title=\"Convert this dataframe to an interactive table.\"\n",
              "            style=\"display:none;\">\n",
              "\n",
              "  <svg xmlns=\"http://www.w3.org/2000/svg\" height=\"24px\" viewBox=\"0 -960 960 960\">\n",
              "    <path d=\"M120-120v-720h720v720H120Zm60-500h600v-160H180v160Zm220 220h160v-160H400v160Zm0 220h160v-160H400v160ZM180-400h160v-160H180v160Zm440 0h160v-160H620v160ZM180-180h160v-160H180v160Zm440 0h160v-160H620v160Z\"/>\n",
              "  </svg>\n",
              "    </button>\n",
              "\n",
              "  <style>\n",
              "    .colab-df-container {\n",
              "      display:flex;\n",
              "      gap: 12px;\n",
              "    }\n",
              "\n",
              "    .colab-df-convert {\n",
              "      background-color: #E8F0FE;\n",
              "      border: none;\n",
              "      border-radius: 50%;\n",
              "      cursor: pointer;\n",
              "      display: none;\n",
              "      fill: #1967D2;\n",
              "      height: 32px;\n",
              "      padding: 0 0 0 0;\n",
              "      width: 32px;\n",
              "    }\n",
              "\n",
              "    .colab-df-convert:hover {\n",
              "      background-color: #E2EBFA;\n",
              "      box-shadow: 0px 1px 2px rgba(60, 64, 67, 0.3), 0px 1px 3px 1px rgba(60, 64, 67, 0.15);\n",
              "      fill: #174EA6;\n",
              "    }\n",
              "\n",
              "    .colab-df-buttons div {\n",
              "      margin-bottom: 4px;\n",
              "    }\n",
              "\n",
              "    [theme=dark] .colab-df-convert {\n",
              "      background-color: #3B4455;\n",
              "      fill: #D2E3FC;\n",
              "    }\n",
              "\n",
              "    [theme=dark] .colab-df-convert:hover {\n",
              "      background-color: #434B5C;\n",
              "      box-shadow: 0px 1px 3px 1px rgba(0, 0, 0, 0.15);\n",
              "      filter: drop-shadow(0px 1px 2px rgba(0, 0, 0, 0.3));\n",
              "      fill: #FFFFFF;\n",
              "    }\n",
              "  </style>\n",
              "\n",
              "    <script>\n",
              "      const buttonEl =\n",
              "        document.querySelector('#df-61a1b2a3-5d60-482b-8fcd-942aee971b4f button.colab-df-convert');\n",
              "      buttonEl.style.display =\n",
              "        google.colab.kernel.accessAllowed ? 'block' : 'none';\n",
              "\n",
              "      async function convertToInteractive(key) {\n",
              "        const element = document.querySelector('#df-61a1b2a3-5d60-482b-8fcd-942aee971b4f');\n",
              "        const dataTable =\n",
              "          await google.colab.kernel.invokeFunction('convertToInteractive',\n",
              "                                                    [key], {});\n",
              "        if (!dataTable) return;\n",
              "\n",
              "        const docLinkHtml = 'Like what you see? Visit the ' +\n",
              "          '<a target=\"_blank\" href=https://colab.research.google.com/notebooks/data_table.ipynb>data table notebook</a>'\n",
              "          + ' to learn more about interactive tables.';\n",
              "        element.innerHTML = '';\n",
              "        dataTable['output_type'] = 'display_data';\n",
              "        await google.colab.output.renderOutput(dataTable, element);\n",
              "        const docLink = document.createElement('div');\n",
              "        docLink.innerHTML = docLinkHtml;\n",
              "        element.appendChild(docLink);\n",
              "      }\n",
              "    </script>\n",
              "  </div>\n",
              "\n",
              "\n",
              "<div id=\"df-07d86c90-8ffd-4def-bb11-8ce28714206e\">\n",
              "  <button class=\"colab-df-quickchart\" onclick=\"quickchart('df-07d86c90-8ffd-4def-bb11-8ce28714206e')\"\n",
              "            title=\"Suggest charts\"\n",
              "            style=\"display:none;\">\n",
              "\n",
              "<svg xmlns=\"http://www.w3.org/2000/svg\" height=\"24px\"viewBox=\"0 0 24 24\"\n",
              "     width=\"24px\">\n",
              "    <g>\n",
              "        <path d=\"M19 3H5c-1.1 0-2 .9-2 2v14c0 1.1.9 2 2 2h14c1.1 0 2-.9 2-2V5c0-1.1-.9-2-2-2zM9 17H7v-7h2v7zm4 0h-2V7h2v10zm4 0h-2v-4h2v4z\"/>\n",
              "    </g>\n",
              "</svg>\n",
              "  </button>\n",
              "\n",
              "<style>\n",
              "  .colab-df-quickchart {\n",
              "      --bg-color: #E8F0FE;\n",
              "      --fill-color: #1967D2;\n",
              "      --hover-bg-color: #E2EBFA;\n",
              "      --hover-fill-color: #174EA6;\n",
              "      --disabled-fill-color: #AAA;\n",
              "      --disabled-bg-color: #DDD;\n",
              "  }\n",
              "\n",
              "  [theme=dark] .colab-df-quickchart {\n",
              "      --bg-color: #3B4455;\n",
              "      --fill-color: #D2E3FC;\n",
              "      --hover-bg-color: #434B5C;\n",
              "      --hover-fill-color: #FFFFFF;\n",
              "      --disabled-bg-color: #3B4455;\n",
              "      --disabled-fill-color: #666;\n",
              "  }\n",
              "\n",
              "  .colab-df-quickchart {\n",
              "    background-color: var(--bg-color);\n",
              "    border: none;\n",
              "    border-radius: 50%;\n",
              "    cursor: pointer;\n",
              "    display: none;\n",
              "    fill: var(--fill-color);\n",
              "    height: 32px;\n",
              "    padding: 0;\n",
              "    width: 32px;\n",
              "  }\n",
              "\n",
              "  .colab-df-quickchart:hover {\n",
              "    background-color: var(--hover-bg-color);\n",
              "    box-shadow: 0 1px 2px rgba(60, 64, 67, 0.3), 0 1px 3px 1px rgba(60, 64, 67, 0.15);\n",
              "    fill: var(--button-hover-fill-color);\n",
              "  }\n",
              "\n",
              "  .colab-df-quickchart-complete:disabled,\n",
              "  .colab-df-quickchart-complete:disabled:hover {\n",
              "    background-color: var(--disabled-bg-color);\n",
              "    fill: var(--disabled-fill-color);\n",
              "    box-shadow: none;\n",
              "  }\n",
              "\n",
              "  .colab-df-spinner {\n",
              "    border: 2px solid var(--fill-color);\n",
              "    border-color: transparent;\n",
              "    border-bottom-color: var(--fill-color);\n",
              "    animation:\n",
              "      spin 1s steps(1) infinite;\n",
              "  }\n",
              "\n",
              "  @keyframes spin {\n",
              "    0% {\n",
              "      border-color: transparent;\n",
              "      border-bottom-color: var(--fill-color);\n",
              "      border-left-color: var(--fill-color);\n",
              "    }\n",
              "    20% {\n",
              "      border-color: transparent;\n",
              "      border-left-color: var(--fill-color);\n",
              "      border-top-color: var(--fill-color);\n",
              "    }\n",
              "    30% {\n",
              "      border-color: transparent;\n",
              "      border-left-color: var(--fill-color);\n",
              "      border-top-color: var(--fill-color);\n",
              "      border-right-color: var(--fill-color);\n",
              "    }\n",
              "    40% {\n",
              "      border-color: transparent;\n",
              "      border-right-color: var(--fill-color);\n",
              "      border-top-color: var(--fill-color);\n",
              "    }\n",
              "    60% {\n",
              "      border-color: transparent;\n",
              "      border-right-color: var(--fill-color);\n",
              "    }\n",
              "    80% {\n",
              "      border-color: transparent;\n",
              "      border-right-color: var(--fill-color);\n",
              "      border-bottom-color: var(--fill-color);\n",
              "    }\n",
              "    90% {\n",
              "      border-color: transparent;\n",
              "      border-bottom-color: var(--fill-color);\n",
              "    }\n",
              "  }\n",
              "</style>\n",
              "\n",
              "  <script>\n",
              "    async function quickchart(key) {\n",
              "      const quickchartButtonEl =\n",
              "        document.querySelector('#' + key + ' button');\n",
              "      quickchartButtonEl.disabled = true;  // To prevent multiple clicks.\n",
              "      quickchartButtonEl.classList.add('colab-df-spinner');\n",
              "      try {\n",
              "        const charts = await google.colab.kernel.invokeFunction(\n",
              "            'suggestCharts', [key], {});\n",
              "      } catch (error) {\n",
              "        console.error('Error during call to suggestCharts:', error);\n",
              "      }\n",
              "      quickchartButtonEl.classList.remove('colab-df-spinner');\n",
              "      quickchartButtonEl.classList.add('colab-df-quickchart-complete');\n",
              "    }\n",
              "    (() => {\n",
              "      let quickchartButtonEl =\n",
              "        document.querySelector('#df-07d86c90-8ffd-4def-bb11-8ce28714206e button');\n",
              "      quickchartButtonEl.style.display =\n",
              "        google.colab.kernel.accessAllowed ? 'block' : 'none';\n",
              "    })();\n",
              "  </script>\n",
              "</div>\n",
              "\n",
              "    </div>\n",
              "  </div>\n"
            ],
            "application/vnd.google.colaboratory.intrinsic+json": {
              "type": "dataframe",
              "variable_name": "df",
              "summary": "{\n  \"name\": \"df\",\n  \"rows\": 962,\n  \"fields\": [\n    {\n      \"column\": \"text\",\n      \"properties\": {\n        \"dtype\": \"string\",\n        \"samples\": [\n          \"FOR a man who spent his career illuminating the vast, dim migrations of people in prehistory, Luca Cavalli-Sforza\\u2019s life was remarkably circular. He first became interested in his major field, genetics, in the house of the geneticist Adriano Buzzati at Belluno, in the hills north of Venice. There he helped to collect thousands of flies in search of mutant Y chromosomes; and though he subsequently travelled the world to study the makeup of its tribes and populations, it was in Belluno that he died.He had vaguely meant to be a doctor, but found his training useless in wartime hospitals. Instead, he took up the genetic study of E. coli bacteria, shutting himself away in the laboratory. But human beings still fascinated him: not least because he increasingly realised that the human genome contained not only the instructions for building future members of the species, but also its entire past. Migration, conquest and isolation had also left their traces in living populations. He set out to find them.The technology available was rudimentary. He pored over names in parish registers and learned to do statistics, convinced that with enough patience he could measure anything. For one early paper he pulled surnames from Italian phone directories as proxies for father-son transmission. By plotting the distribution of names in 91 provinces, he showed that the Apennines\\u2014the backbone of Italy\\u2014had formed a natural barrier to migration for millennia, resulting in genetically distinct populations on either side. And because he could also measure variation in blood proteins, such as that between A, B and O blood types, he also collected blood samples from townsfolk all over the Parma valley to show that marriage between related individuals had led to many of the genetic differences between those towns.For two decades from 1970, when he took up a professorship at Stanford, he devoted himself to this subject. By the 1990s he was able to study variation in DNA itself. When he and his colleagues instructed a computer to sort around 1,000 people from across the globe into five clusters by similarity of DNA, the clusters matched the labels by which humans had long grouped themselves intuitively: West Eurasians, East Asians, Native Americans, New Guineans and Africans. He soon concluded, though, that race was not a scientifically valid way to classify them. Europeans, for example, were about two-thirds Asian and one-third African, but after millennia of mixing there was no such thing as pure Asian or pure African either. Skin colour, or the shape of a nose, were just superficial adaptations to climate and place.He represented his genetic data as \\u201ctrees\\u201d branching over time: simplistic diagrams, but beautiful in their simplicity. They often agreed, as he had hoped, with the findings of linguists and archaeologists: suggesting, for example, that humanity arose in Africa, where it stayed for a long time before moving outwards, somewhere between 100,000 and 50,000 years ago. The earliest African migrants probably reached Asia first, moving on to Oceania, Europe and America, in that order.He particularly liked to tell his story of farmers. By checking variations between individuals based on their blood groups, he discerned a gradient in that variation that stretched south-east-to-north-west across Europe. This he saw as the genetic signature of farming after its invention around 9,000 years ago in the Fertile Crescent. As early farmers radiated out of the Near East, he argued, they bred with indigenous hunter-gatherers, until by the time they reached colder climates their genes were a good mix of both.This theory stood until the 2000s, when new technology reversed it. He had often carped about the public\\u2019s interest in fossils, so much less informative\\u2014to him\\u2014than genetic studies of the living. Once DNA could be extracted from ancient bones, however, it showed that although agriculture had indeed reached Europe from the Middle East, a wave of immigration from the north-east, starting about 5,000 years ago, had diluted those first farmers with tall herdsmen from the Pontic Steppe. People had moved and mixed in prehistory more than he thought; and not all ancient events, as he supposed, had left their mark in modern populations.Even more controversial was his Human Genome Diversity Project, which he set up in the 1990s. He wanted to study isolated populations in order to understand where all the others came from; to measure the background of drift, or genetic change, that takes place without marauding or migration. Some people thought this racist, and a more worldly man might have realised that. A few critics even brought up his membership of Benito Mussolini\\u2019s fascist youth organisation, compulsory before the war. All this dismayed him, as he had done so much to strike down \\u201cscientific\\u201d studies based on race. The project died in its original form, though it was resurrected as the Genographic Project, which goes on.From Belluno he witnessed the overturning of many other conclusions. Every evolutionary story he had touched became more complex by the day. Yet he could comfort himself that without his original vision for the study of human history through its genes, much of that great debate would not have happened at all.\",\n          \"T HE FIRST outbreak of a novel disease is the opening scene of a whodunnit. In 1976, when more than two dozen members of the American Legion died after a convention in Philadelphia, public-health officials spent months scouring the hotel they had met in before finally tracking down the culprit in the water tank on the roof: a new bacterium which, having caused the first known cases of Legionnaires\\u2019 disease, was named Legionella. In the 1980s it took years of hard work and acrimonious argument among epidemiologists and virologists to blame the terrible and varied symptoms of AIDS on HIV, a virus of a type never previously seen in humans.For covid-19, the mystery was solved almost as soon as it had begun. The novel pneumonia that doctors in Wuhan noticed in December 2019 immediately brought to mind Severe Acute Respiratory Syndrome (SARS), a disease caused by a coronavirus which broke out in 2002. As a result of SARS and the subsequent outbreak in 2012 of Middle East Respiratory Syndrome (MERS), also caused by a coronavirus, there were already established protocols for growing cells from the lining of the nose, throat and lungs in order to look for coronavirus infection. They were soon put to use.To identify the possible coronavirus responsible meant producing a sequence of its genome. The first step in this process was to extract RNA\\u2014the molecule on which coronavirus genomes are written\\u2014from the cell cultures. The genetic sequences in these RNA molecules then had to be transcribed into complementary bits of DNA, because that is what automated sequencing machines work with (see part A of diagram).Computer programs assembled the sequences those machines produced into a recognisable, if gappy, coronavirus genome. Researchers used this to make DNA \\u201cprimers\\u201d with which to fish out the not-yet-sequenced bits of the genome. Finished sequences were published less than two weeks after the process had begun. On January 12th 2020 the world knew its enemy\\u2014soon thereafter named SARS-CoV-2\\u2014down to the last letter of its genome.In terms of the science done, this was all routine; the appropriate use of standard laboratory techniques. In terms of its impact, it was enormous. Knowing the viral sequence was fundamental to vaccination efforts, made it possible to track the virus\\u2019s evolution and, most immediately, made it possible to test people with a cough and see if they were infected. The first AIDS tests were not available until four years after medical science became aware of the condition they tested for. For SARS it took six months. Procedures for testing swabs from the nose and throat for RNA from SARS-CoV-2 were published 11 days after the genome sequence, on January 23rd.There was, however, a drawback to the tests. They required suitably equipped laboratories. Most countries did not have nearly enough of the relevant lab capacity; in others, much of it was being used for different things. \\u201cPrior to March 23rd my lab had never performed a viral diagnostic,\\u201d says Stacey Gabriel, who runs genetic-sequencing operations at the Broad Institute in Cambridge, Massachusetts. But with public institutions swamped she and her colleagues created one of the largest testing shops on America\\u2019s east coast from scratch, reconfiguring the specialised robots that populate one of the world\\u2019s most advanced cancer-genetics labs to do the grunt work.Dr Gabriel learned two lessons in the process. The first is that uniformity matters a lot. The Broad started off testing samples from Massachusetts nursing homes which came in containers of varying size and with various amounts of liquid, some accompanied by handwritten forms, some by barcodes. Dealing with such messiness is no task for a robot, and so to begin with just a few thousand samples a day passed through machines capable of handling far more. The second is that you need software to track the whole process. Commercial software, dry swabs and barcoding soon had the lab firing on all cylinders. By March 2021 it could handle 200,000 tests a day and was serving customers as far afield as New York.The data such labs produce are not just for patients and doctors. In most countries covid-19 is a notifiable disease; the authorities have a legal right to know who has been found to be infected. When a sample tests positive the lab has to pass the identity of the person it came from on to public-health officials. At that point a new sort of detective work begins: where did that person\\u2014the index case, in public-health speak\\u2014pick up the virus? To whom might they have given it?Several East Asian countries demonstrated that, if started in the earliest days of an epidemic and pursued with vigour and persistence, such contact tracing can be a powerful tool. Some, such as Singapore and Taiwan, benefited in this from their experience with SARS in the mid-2000s; tracing systems set in place back then were put to use with an urgency born of experience. A level of invasiveness from which most Western authorities shied away was often employed. In South Korea, for instance, contact tracers were able to download a list of all financial transactions made by those who tested positive; they could then obtain CCTV footage from shops the index case had visited in search of other customers to check up on. \\u201c[Such snooping] has come up with discussions I\\u2019ve had with policymakers,\\u201d says Christophe Fraser, a digital epidemiologist at the University of Oxford. \\u201cWe got very hung up on the idea of contact tracing disrupting people\\u2019s lives.\\u201dWestern governments acted in a slower, less thoroughgoing way. They failed to track the initial spread, in part because of insufficient testing capacity (and, in America, dud tests from the Centres for Disease Control and Prevention). They ended up with much less impressive systems. Tom Frieden, a former director of the CDC, says he thinks that, at its current level of effort, America could plausibly trace about 15,000 cases a day\\u2014a level that has been handsomely exceeded every day since April 2nd 2020. Britain has earmarked \\u00a337bn for testing and tracing over the 2020 and 2021 financial years, and though it may well spend less, the shoddiness that has dogged some elements of the campaign, such as a database cock-up which lost thousands of results in September 2020, will be remembered.A basic problem is that contact tracing is a lot of work. Theoretical assessments based on analysing social networks and experience in Asia both suggest that some 30 contacts need to be identified for each index case. Digital tools can lessen the load. Resolve to Save Lives, a campaign run by Vital Strategies, an NGO, developed Locator, which taps into credit data to help tracers track down people who may have caught the virus from a specific index case. But the amount of work required remained enormous.A much discussed alternative to such programmes was to use the world\\u2019s most ubiquitous tracking devices: smartphones. Google and Apple worked together to develop a system which enabled phones to keep a list of occasions when they were near another phone for a significant period of time, and the identity of that second phone. When someone tests positive for SARS-CoV-2 they are asked to send a message from their phone which is used to notify all the other phones they had been near within a particular window of time. But everything is peer-to-peer; neither the big tech companies nor the public-health authorities get a list of contacts.Unfortunately this built-in privacy makes it hard to assess the technology\\u2019s efficacy. British and Swiss studies suggest such apps do reduce spread, but not enough to make them more than an also-ran technology. None of the successful contact-tracing systems in East Asia relies on such things to any significant extent.Even without good tracing, self isolation of those who tested positive helped slow the spread of the disease. But over time the shortcomings of the initial testing technology, reverse-transcriptase polymerase chain reaction (RT-PCR), became ever more apparent. It is conceptually elegant (see diagram) and easy for labs to use. But despite its familiarity, reliability and sensitivity, it has real disadvantages.One is that it needs labs, and is carried out most efficiently in big ones. This means samples may have to travel a long way. It also means that they can get held in queues. The Broad runs its RT-PCR tests in just three and a half hours; but the average sample takes 15 hours to process. Results from RT-PCR tests normally come in days not minutes.Another problem is that though the presence of viral RNA clearly shows that a person has been infected, it says very little about where they stand in the course of the disease; RNA is detectible from very soon after infection to long after the disease has run its course. In public-health terms what is needed is a test that spots people who are actually infectious\\u2014people with cells in their noses and throats actively churning out virus particles.To look for the SARS-CoV-2 particles themselves means looking for their distinctive protein components, not for the RNA that tells cells how to make them. The most detectible such component is the spike protein which studs the particles\\u2019 outer membranes. And one of the basic rules of modern biotechnology is that when you want to find a protein, use an antibody.Antibodies are large molecules that come in millions of varieties, each of which sticks to one target\\u2014known as that antibody\\u2019s \\u201cantigen\\u201d\\u2014and one target only. A handy technology called lateral-flow testing makes use of that specificity. A sample is placed at one end of a porous membrane and, as it seeps along to the other, encounters a line of antibodies designed to recognise it. When the sample is urine and the antigen is a hormone found in expectant women, you have a pregnancy test. When the sample is mucus from a swab and the antigen is the spike protein, it is a covid-19 test\\u2014a cheap, convenient one which can provide results with in half an hour. Such tests may not pick up all the people in whom RT-PCR might detect a trace of the virus. But if their nose and throat cells are not producing enough antigen for the test to detect, they are probably not producing enough to be infectious, either.At the beginning of the epidemic \\u201cthe supply chain for the lateral-flow tests wasn\\u2019t there,\\u201d says Dr Gabriel. Chris Hand, the chairman of Abingdon Health, a British contract manufacturer of lateral-flow tests, says the main bottleneck was the speciality membranes that are part of every test kit. \\u201cThey come on large reels of 100 metres plus, which go through automated equipment to add biochemicals by spraying them at low volumes,\\u201d he says. But once the biochemicals\\u2014the bespoke antibodies and some more generic bits and bobs\\u2014are ready, the production processes in place and the packaging sorted, the tests could be churned out by the million.New technologies now reaching the market will further change the dynamics of test and trace. QuantuMDx is one of a number of companies developing automated PCR-in-a-box systems that provide results within a couple of hours. Jonathan O\\u2019Halloran, the firm\\u2019s boss, says the British company has been relying on its own testing system for the past 22 weeks, testing its 90 staff members every morning (they are free to decline). About once a week lunchtime brings the news that someone has tested positive; they are immediately sent home to isolate. When things are done this fast, the fact that RNA is detectable before people are infectious is a plus; isolating on the basis of an early PCR test means no one ever turns up to work infectious. The company claims not to have lost a single day to infection.A combination of local, automated PCR and lateral-flow tests could be the basis of an ideal testing system\\u2014one which has a chance of keeping ahead of, and containing, a low level of disease rather than lagging behind one that is shooting up. Antigen tests would be used to scan the population for new infections. Those found would be referred to contact tracers; contacts who might have been infected could then be PCR-tested to find out which of them actually were.Great for public health. No benefit, in itself, to the index cases who would wait, isolated, to see what fate the virus and their immune systems had in store for them\\u2014painfully aware that their next encounter with the wonders of modern medical technology could be in a hospital bed. \\u25a0\",\n          \"FOR decades scientists have been trying to understand schizophrenia, a distressing disorder that afflicts one in a hundred people. Its destructive symptoms include hallucinations, delusions, muddled thoughts and changes in behaviour. The best drugs offer little more than the ability to target symptoms. There is presently no hope of a cure because its ultimate cause has long been a mystery.Schizophrenia is known to run in families, so it has always been thought that genes might shed light on the origins of the disease. But the genetics have proved harder to unpick than anyone imagined. It is only now, 16 years after the human genome was first sequenced, that scientists have homed in on the relevant genes.For more than five years researchers led by a group at the Broad Institute in Cambridge, Massachusetts, collected more than 100,000 DNA samples from around the world. They were looking for regions of the human genome that might be harbouring variants that increased the risk of schizophrenia. What they found implicated more than a hundred genes and provided a strong pointer towards a portion of the genome associated with infectious diseases. Many of these genes did not operate independently of each other, and further detective work revealed that the most important source of variation lay within a particular gene known as C4.Steven McCarroll, of Harvard Medical School, and his colleagues involved in the study, report in Nature this week that particular forms of the C4 gene led to a greater risk of developing schizophrenia. What makes C4 hard to identify is that it comes in many different forms and has an intriguing function. The gene produces a protein known as a complement factor, which helps to tag pathogens for removal by the body\\u2019s immune system. In the brain, however, it appears to be tied to a developmental process known as \\u201csynaptic pruning\\u201d, in which neurons are continuously eliminated throughout childhood and into early adulthood.In other words schizophrenics may be producing too much of a protein that creates a signal that synaptic connections should be removed. During late adolescence the brain undergoes widespread synaptic pruning, particularly in the cerebral cortex. This is also the age at which schizophrenia symptoms tend to begin.David Goldstein, director of the Institute for Genomic Medicine at Columbia University, says he is persuaded by the research, but reckons it is important for the medical community to examine the study carefully because the region of the genome under discussion is hard to work on and is \\u201cstuffed full of genetic variation\\u201d. That makes it difficult to point to a single gene and say \\u201cit\\u2019s this\\u201d, he adds.As to whether those with a family history of schizophrenia ought to have their genome sequenced to assess their susceptibility, Dr Goldstein is keen to point out that more work needs to be done on the risks that different forms of the genes bring. Moreover, other factors can be implicated in developing schizophrenia, ranging from taking psychoactive drugs to being exposed to viruses while in the womb.Synaptic pruning is so crucial in brain development researchers wonder what other mental disorders might be tied to its aberrant activation or regulation. This study is bound to initiate a flurry of similar work. Alzheimer\\u2019s disease, for one, has been tied to, among other things, excessive destruction of synapses. And it is already believed that autistic individuals have a surplus of synapses in the brain, due to a slowdown in normal brain pruning. Indeed, the onset of autism often follows a period when there is a burst of synapse formation. One big step forward in schizophrenia research could have an enormous impact on the way other neurological disorders are viewed and treated.\"\n        ],\n        \"num_unique_values\": 961,\n        \"semantic_type\": \"\",\n        \"description\": \"\"\n      }\n    },\n    {\n      \"column\": \"title\",\n      \"properties\": {\n        \"dtype\": \"string\",\n        \"samples\": [\n          \"Genetic engineering may help control disease-carrying mosquitoes\",\n          \"Covid-19 has many faces\",\n          \"Studying broken chromosomes can illuminate neuroscience\"\n        ],\n        \"num_unique_values\": 962,\n        \"semantic_type\": \"\",\n        \"description\": \"\"\n      }\n    },\n    {\n      \"column\": \"date\",\n      \"properties\": {\n        \"dtype\": \"object\",\n        \"min\": \" Apr 10th 1997\",\n        \"max\": \" Sep 9th 2020\",\n        \"samples\": [\n          \" Aug 29th 2019\",\n          \" Aug 24th 2006\",\n          \" Jan 11th 2022\"\n        ],\n        \"num_unique_values\": 783,\n        \"semantic_type\": \"\",\n        \"description\": \"\"\n      }\n    },\n    {\n      \"column\": \"link\",\n      \"properties\": {\n        \"dtype\": \"string\",\n        \"samples\": [\n          \"https://www.economist.com/science-and-technology/2021/05/26/genetic-engineering-may-help-control-disease-carrying-mosquitoes\",\n          \"https://www.economist.com/science-and-technology/2020/05/09/covid-19-has-many-faces\",\n          \"https://www.economist.com/science-and-technology/2023/03/08/studying-broken-chromosomes-can-illuminate-neuroscience\"\n        ],\n        \"num_unique_values\": 962,\n        \"semantic_type\": \"\",\n        \"description\": \"\"\n      }\n    },\n    {\n      \"column\": \"cleaned_text\",\n      \"properties\": {\n        \"dtype\": \"string\",\n        \"samples\": [\n          \"for a man who spent his career illuminating the vast dim migrations of people in prehistory luca cavallisforza s life was remarkably circular he first became interested in his major field genetics in the house of the geneticist adriano buzzati at belluno in the hills north of venice there he helped to collect thousands of flies in search of mutant y chromosomes  and though he subsequently travelled the world to study the makeup of its tribes and populations it was in belluno that he diedhe had vaguely meant to be a doctor but found his training useless in wartime hospitals instead he took up the genetic study of e coli bacteria shutting himself away in the laboratory but human beings still fascinated him  not least because he increasingly realised that the human genome contained not only the instructions for building future members of the species but also its entire past migration conquest and isolation had also left their traces in living populations he set out to find themthe technology available was rudimentary he pored over names in parish registers and learned to do statistics convinced that with enough patience he could measure anything for one early paper he pulled surnames from italian phone directories as proxies for fatherson transmission by plotting the distribution of names in 91 provinces he showed that the apennines the backbone of italy had formed a natural barrier to migration for millennia resulting in genetically distinct populations on either side and because he could also measure variation in blood proteins such as that between a b and o blood types he also collected blood samples from townsfolk all over the parma valley to show that marriage between related individuals had led to many of the genetic differences between those townsfor two decades from 1970 when he took up a professorship at stanford he devoted himself to this subject by the 1990s he was able to study variation in dna itself when he and his colleagues instructed a computer to sort around 1000 people from across the globe into five clusters by similarity of dna the clusters matched the labels by which humans had long grouped themselves intuitively  west eurasians east asians native americans new guineans and africans he soon concluded though that race was not a scientifically valid way to classify them europeans for example were about twothirds asian and onethird african but after millennia of mixing there was no such thing as pure asian or pure african either skin colour or the shape of a nose were just superficial adaptations to climate and placehe represented his genetic data as  trees  branching over time  simplistic diagrams but beautiful in their simplicity they often agreed as he had hoped with the findings of linguists and archaeologists  suggesting for example that humanity arose in africa where it stayed for a long time before moving outwards somewhere between 100000 and 50000 years ago the earliest african migrants probably reached asia first moving on to oceania europe and america in that orderhe particularly liked to tell his story of farmers by checking variations between individuals based on their blood groups he discerned a gradient in that variation that stretched southeasttonorthwest across europe this he saw as the genetic signature of farming after its invention around 9000 years ago in the fertile crescent as early farmers radiated out of the near east he argued they bred with indigenous huntergatherers until by the time they reached colder climates their genes were a good mix of boththis theory stood until the 2000s when new technology reversed it he had often carped about the public s interest in fossils so much less informative to him than genetic studies of the living once dna could be extracted from ancient bones however it showed that although agriculture had indeed reached europe from the middle east a wave of immigration from the northeast starting about 5000 years ago had diluted those first farmers with tall herdsmen from the pontic steppe people had moved and mixed in prehistory more than he thought  and not all ancient events as he supposed had left their mark in modern populationseven more controversial was his human genome diversity project which he set up in the 1990s he wanted to study isolated populations in order to understand where all the others came from  to measure the background of drift or genetic change that takes place without marauding or migration some people thought this racist and a more worldly man might have realised that a few critics even brought up his membership of benito mussolini s fascist youth organisation compulsory before the war all this dismayed him as he had done so much to strike down  scientific  studies based on race the project died in its original form though it was resurrected as the genographic project which goes onfrom belluno he witnessed the overturning of many other conclusions every evolutionary story he had touched became more complex by the day yet he could comfort himself that without his original vision for the study of human history through its genes much of that great debate would not have happened at all\",\n          \"t he first outbreak of a novel disease is the opening scene of a whodunnit in 1976 when more than two dozen members of the american legion died after a convention in philadelphia publichealth officials spent months scouring the hotel they had met in before finally tracking down the culprit in the water tank on the roof  a new bacterium which having caused the first known cases of legionnaires  disease was named legionella in the 1980s it took years of hard work and acrimonious argument among epidemiologists and virologists to blame the terrible and varied symptoms of aids on hiv a virus of a type never previously seen in humansfor covid19 the mystery was solved almost as soon as it had begun the novel pneumonia that doctors in wuhan noticed in december 2019 immediately brought to mind severe acute respiratory syndrome sars a disease caused by a coronavirus which broke out in 2002 as a result of sars and the subsequent outbreak in 2012 of middle east respiratory syndrome mers also caused by a coronavirus there were already established protocols for growing cells from the lining of the nose throat and lungs in order to look for coronavirus infection they were soon put to useto identify the possible coronavirus responsible meant producing a sequence of its genome the first step in this process was to extract rna the molecule on which coronavirus genomes are written from the cell cultures the genetic sequences in these rna molecules then had to be transcribed into complementary bits of dna because that is what automated sequencing machines work with see part a of diagramcomputer programs assembled the sequences those machines produced into a recognisable if gappy coronavirus genome researchers used this to make dna  primers  with which to fish out the notyetsequenced bits of the genome finished sequences were published less than two weeks after the process had begun on january 12th 2020 the world knew its enemy soon thereafter named sarscov2 down to the last letter of its genomein terms of the science done this was all routine  the appropriate use of standard laboratory techniques in terms of its impact it was enormous knowing the viral sequence was fundamental to vaccination efforts made it possible to track the virus s evolution and most immediately made it possible to test people with a cough and see if they were infected the first aids tests were not available until four years after medical science became aware of the condition they tested for for sars it took six months procedures for testing swabs from the nose and throat for rna from sarscov2 were published 11 days after the genome sequence on january 23rdthere was however a drawback to the tests they required suitably equipped laboratories most countries did not have nearly enough of the relevant lab capacity  in others much of it was being used for different things  prior to march 23rd my lab had never performed a viral diagnostic  says stacey gabriel who runs geneticsequencing operations at the broad institute in cambridge massachusetts but with public institutions swamped she and her colleagues created one of the largest testing shops on america s east coast from scratch reconfiguring the specialised robots that populate one of the world s most advanced cancergenetics labs to do the grunt workdr gabriel learned two lessons in the process the first is that uniformity matters a lot the broad started off testing samples from massachusetts nursing homes which came in containers of varying size and with various amounts of liquid some accompanied by handwritten forms some by barcodes dealing with such messiness is no task for a robot and so to begin with just a few thousand samples a day passed through machines capable of handling far more the second is that you need software to track the whole process commercial software dry swabs and barcoding soon had the lab firing on all cylinders by march 2021 it could handle 200000 tests a day and was serving customers as far afield as new yorkthe data such labs produce are not just for patients and doctors in most countries covid19 is a notifiable disease  the authorities have a legal right to know who has been found to be infected when a sample tests positive the lab has to pass the identity of the person it came from on to publichealth officials at that point a new sort of detective work begins  where did that person the index case in publichealth speak pick up the virus to whom might they have given itseveral east asian countries demonstrated that if started in the earliest days of an epidemic and pursued with vigour and persistence such contact tracing can be a powerful tool some such as singapore and taiwan benefited in this from their experience with sars in the mid2000s  tracing systems set in place back then were put to use with an urgency born of experience a level of invasiveness from which most western authorities shied away was often employed in south korea for instance contact tracers were able to download a list of all financial transactions made by those who tested positive  they could then obtain cctv footage from shops the index case had visited in search of other customers to check up on   such snooping  has come up with discussions i ve had with policymakers  says christophe fraser a digital epidemiologist at the university of oxford  we got very hung up on the idea of contact tracing disrupting people s lives western governments acted in a slower less thoroughgoing way they failed to track the initial spread in part because of insufficient testing capacity and in america dud tests from the centres for disease control and prevention they ended up with much less impressive systems tom frieden a former director of the cdc says he thinks that at its current level of effort america could plausibly trace about 15000 cases a day a level that has been handsomely exceeded every day since april 2nd 2020 britain has earmarked  37bn for testing and tracing over the 2020 and 2021 financial years and though it may well spend less the shoddiness that has dogged some elements of the campaign such as a database cockup which lost thousands of results in september 2020 will be remembereda basic problem is that contact tracing is a lot of work theoretical assessments based on analysing social networks and experience in asia both suggest that some 30 contacts need to be identified for each index case digital tools can lessen the load resolve to save lives a campaign run by vital strategies an ngo developed locator which taps into credit data to help tracers track down people who may have caught the virus from a specific index case but the amount of work required remained enormousa much discussed alternative to such programmes was to use the world s most ubiquitous tracking devices  smartphones google and apple worked together to develop a system which enabled phones to keep a list of occasions when they were near another phone for a significant period of time and the identity of that second phone when someone tests positive for sarscov2 they are asked to send a message from their phone which is used to notify all the other phones they had been near within a particular window of time but everything is peertopeer  neither the big tech companies nor the publichealth authorities get a list of contactsunfortunately this builtin privacy makes it hard to assess the technology s efficacy british and swiss studies suggest such apps do reduce spread but not enough to make them more than an alsoran technology none of the successful contacttracing systems in east asia relies on such things to any significant extenteven without good tracing self isolation of those who tested positive helped slow the spread of the disease but over time the shortcomings of the initial testing technology reversetranscriptase polymerase chain reaction rtpcr became ever more apparent it is conceptually elegant see diagram and easy for labs to use but despite its familiarity reliability and sensitivity it has real disadvantagesone is that it needs labs and is carried out most efficiently in big ones this means samples may have to travel a long way it also means that they can get held in queues the broad runs its rtpcr tests in just three and a half hours  but the average sample takes 15 hours to process results from rtpcr tests normally come in days not minutesanother problem is that though the presence of viral rna clearly shows that a person has been infected it says very little about where they stand in the course of the disease  rna is detectible from very soon after infection to long after the disease has run its course in publichealth terms what is needed is a test that spots people who are actually infectious people with cells in their noses and throats actively churning out virus particlesto look for the sarscov2 particles themselves means looking for their distinctive protein components not for the rna that tells cells how to make them the most detectible such component is the spike protein which studs the particles  outer membranes and one of the basic rules of modern biotechnology is that when you want to find a protein use an antibodyantibodies are large molecules that come in millions of varieties each of which sticks to one target known as that antibody s  antigen  and one target only a handy technology called lateralflow testing makes use of that specificity a sample is placed at one end of a porous membrane and as it seeps along to the other encounters a line of antibodies designed to recognise it when the sample is urine and the antigen is a hormone found in expectant women you have a pregnancy test when the sample is mucus from a swab and the antigen is the spike protein it is a covid19 test a cheap convenient one which can provide results with in half an hour such tests may not pick up all the people in whom rtpcr might detect a trace of the virus but if their nose and throat cells are not producing enough antigen for the test to detect they are probably not producing enough to be infectious eitherat the beginning of the epidemic  the supply chain for the lateralflow tests wasn t there  says dr gabriel chris hand the chairman of abingdon health a british contract manufacturer of lateralflow tests says the main bottleneck was the speciality membranes that are part of every test kit  they come on large reels of 100 metres plus which go through automated equipment to add biochemicals by spraying them at low volumes  he says but once the biochemicals the bespoke antibodies and some more generic bits and bobs are ready the production processes in place and the packaging sorted the tests could be churned out by the millionnew technologies now reaching the market will further change the dynamics of test and trace quantumdx is one of a number of companies developing automated pcrinabox systems that provide results within a couple of hours jonathan o halloran the firm s boss says the british company has been relying on its own testing system for the past 22 weeks testing its 90 staff members every morning they are free to decline about once a week lunchtime brings the news that someone has tested positive  they are immediately sent home to isolate when things are done this fast the fact that rna is detectable before people are infectious is a plus  isolating on the basis of an early pcr test means no one ever turns up to work infectious the company claims not to have lost a single day to infectiona combination of local automated pcr and lateralflow tests could be the basis of an ideal testing system one which has a chance of keeping ahead of and containing a low level of disease rather than lagging behind one that is shooting up antigen tests would be used to scan the population for new infections those found would be referred to contact tracers  contacts who might have been infected could then be pcrtested to find out which of them actually weregreat for public health no benefit in itself to the index cases who would wait isolated to see what fate the virus and their immune systems had in store for them painfully aware that their next encounter with the wonders of modern medical technology could be in a hospital bed  \",\n          \"for decades scientists have been trying to understand schizophrenia a distressing disorder that afflicts one in a hundred people its destructive symptoms include hallucinations delusions muddled thoughts and changes in behaviour the best drugs offer little more than the ability to target symptoms there is presently no hope of a cure because its ultimate cause has long been a mysteryschizophrenia is known to run in families so it has always been thought that genes might shed light on the origins of the disease but the genetics have proved harder to unpick than anyone imagined it is only now 16 years after the human genome was first sequenced that scientists have homed in on the relevant genesfor more than five years researchers led by a group at the broad institute in cambridge massachusetts collected more than 100000 dna samples from around the world they were looking for regions of the human genome that might be harbouring variants that increased the risk of schizophrenia what they found implicated more than a hundred genes and provided a strong pointer towards a portion of the genome associated with infectious diseases many of these genes did not operate independently of each other and further detective work revealed that the most important source of variation lay within a particular gene known as c4steven mccarroll of harvard medical school and his colleagues involved in the study report in nature this week that particular forms of the c4 gene led to a greater risk of developing schizophrenia what makes c4 hard to identify is that it comes in many different forms and has an intriguing function the gene produces a protein known as a complement factor which helps to tag pathogens for removal by the body s immune system in the brain however it appears to be tied to a developmental process known as  synaptic pruning  in which neurons are continuously eliminated throughout childhood and into early adulthoodin other words schizophrenics may be producing too much of a protein that creates a signal that synaptic connections should be removed during late adolescence the brain undergoes widespread synaptic pruning particularly in the cerebral cortex this is also the age at which schizophrenia symptoms tend to begindavid goldstein director of the institute for genomic medicine at columbia university says he is persuaded by the research but reckons it is important for the medical community to examine the study carefully because the region of the genome under discussion is hard to work on and is  stuffed full of genetic variation  that makes it difficult to point to a single gene and say  it s this  he addsas to whether those with a family history of schizophrenia ought to have their genome sequenced to assess their susceptibility dr goldstein is keen to point out that more work needs to be done on the risks that different forms of the genes bring moreover other factors can be implicated in developing schizophrenia ranging from taking psychoactive drugs to being exposed to viruses while in the wombsynaptic pruning is so crucial in brain development researchers wonder what other mental disorders might be tied to its aberrant activation or regulation this study is bound to initiate a flurry of similar work alzheimer s disease for one has been tied to among other things excessive destruction of synapses and it is already believed that autistic individuals have a surplus of synapses in the brain due to a slowdown in normal brain pruning indeed the onset of autism often follows a period when there is a burst of synapse formation one big step forward in schizophrenia research could have an enormous impact on the way other neurological disorders are viewed and treated\"\n        ],\n        \"num_unique_values\": 961,\n        \"semantic_type\": \"\",\n        \"description\": \"\"\n      }\n    },\n    {\n      \"column\": \"fully_cleaned_text\",\n      \"properties\": {\n        \"dtype\": \"string\",\n        \"samples\": [\n          \"man spent career illuminating vast dim migrations people prehistory luca cavallisforza life remarkably circular first became interested major field genetics house geneticist adriano buzzati belluno hills north venice helped collect thousands flies search mutant chromosomes though subsequently travelled world study makeup tribes populations belluno diedhe vaguely meant doctor found training useless wartime hospitals instead took genetic study e coli bacteria shutting away laboratory human beings still fascinated least increasingly realised human genome contained instructions building future members species also entire past migration conquest isolation also left traces living populations set find themthe technology available rudimentary pored names parish registers learned statistics convinced enough patience could measure anything one early paper pulled surnames italian phone directories proxies fatherson transmission plotting distribution names 91 provinces showed apennines backbone italy formed natural barrier migration millennia resulting genetically distinct populations either side could also measure variation blood proteins b blood types also collected blood samples townsfolk parma valley show marriage related individuals led many genetic differences townsfor two decades 1970 took professorship stanford devoted subject 1990s able study variation dna colleagues instructed computer sort around 1000 people across globe five clusters similarity dna clusters matched labels humans long grouped intuitively west eurasians east asians native americans new guineans africans soon concluded though race scientifically valid way classify europeans example twothirds asian onethird african millennia mixing thing pure asian pure african either skin colour shape nose superficial adaptations climate placehe represented genetic data trees branching time simplistic diagrams beautiful simplicity often agreed hoped findings linguists archaeologists suggesting example humanity arose africa stayed long time moving outwards somewhere 100000 50000 years ago earliest african migrants probably reached asia first moving oceania europe america orderhe particularly liked tell story farmers checking variations individuals based blood groups discerned gradient variation stretched southeasttonorthwest across europe saw genetic signature farming invention around 9000 years ago fertile crescent early farmers radiated near east argued bred indigenous huntergatherers time reached colder climates genes good mix boththis theory stood 2000s new technology reversed often carped public interest fossils much less informative genetic studies living dna could extracted ancient bones however showed although agriculture indeed reached europe middle east wave immigration northeast starting 5000 years ago diluted first farmers tall herdsmen pontic steppe people moved mixed prehistory thought ancient events supposed left mark modern populationseven controversial human genome diversity project set 1990s wanted study isolated populations order understand others came measure background drift genetic change takes place without marauding migration people thought racist worldly man might realised critics even brought membership benito mussolini fascist youth organisation compulsory war dismayed done much strike scientific studies based race project died original form though resurrected genographic project goes onfrom belluno witnessed overturning many conclusions every evolutionary story touched became complex day yet could comfort without original vision study human history genes much great debate would happened\",\n          \"first outbreak novel disease opening scene whodunnit 1976 two dozen members american legion died convention philadelphia publichealth officials spent months scouring hotel met finally tracking culprit water tank roof new bacterium caused first known cases legionnaires disease named legionella 1980s took years hard work acrimonious argument among epidemiologists virologists blame terrible varied symptoms aids hiv virus type never previously seen humansfor covid19 mystery solved almost soon begun novel pneumonia doctors wuhan noticed december 2019 immediately brought mind severe acute respiratory syndrome sars disease caused coronavirus broke 2002 result sars subsequent outbreak 2012 middle east respiratory syndrome mers also caused coronavirus already established protocols growing cells lining nose throat lungs order look coronavirus infection soon put useto identify possible coronavirus responsible meant producing sequence genome first step process extract rna molecule coronavirus genomes written cell cultures genetic sequences rna molecules transcribed complementary bits dna automated sequencing machines work see part diagramcomputer programs assembled sequences machines produced recognisable gappy coronavirus genome researchers used make dna primers fish notyetsequenced bits genome finished sequences published less two weeks process begun january 12th 2020 world knew enemy soon thereafter named sarscov2 last letter genomein terms science done routine appropriate use standard laboratory techniques terms impact enormous knowing viral sequence fundamental vaccination efforts made possible track virus evolution immediately made possible test people cough see infected first aids tests available four years medical science became aware condition tested sars took six months procedures testing swabs nose throat rna sarscov2 published 11 days genome sequence january 23rdthere however drawback tests required suitably equipped laboratories countries nearly enough relevant lab capacity others much used different things prior march 23rd lab never performed viral diagnostic says stacey gabriel runs geneticsequencing operations broad institute cambridge massachusetts public institutions swamped colleagues created one largest testing shops america east coast scratch reconfiguring specialised robots populate one world advanced cancergenetics labs grunt workdr gabriel learned two lessons process first uniformity matters lot broad started testing samples massachusetts nursing homes came containers varying size various amounts liquid accompanied handwritten forms barcodes dealing messiness task robot begin thousand samples day passed machines capable handling far second need software track whole process commercial software dry swabs barcoding soon lab firing cylinders march 2021 could handle 200000 tests day serving customers far afield new yorkthe data labs produce patients doctors countries covid19 notifiable disease authorities legal right know found infected sample tests positive lab pass identity person came publichealth officials point new sort detective work begins person index case publichealth speak pick virus might given itseveral east asian countries demonstrated started earliest days epidemic pursued vigour persistence contact tracing powerful tool singapore taiwan benefited experience sars mid2000s tracing systems set place back put use urgency born experience level invasiveness western authorities shied away often employed south korea instance contact tracers able download list financial transactions made tested positive could obtain cctv footage shops index case visited search customers check snooping come discussions policymakers says christophe fraser digital epidemiologist university oxford got hung idea contact tracing disrupting people lives western governments acted slower less thoroughgoing way failed track initial spread part insufficient testing capacity america dud tests centres disease control prevention ended much less impressive systems tom frieden former director cdc says thinks current level effort america could plausibly trace 15000 cases day level handsomely exceeded every day since april 2nd 2020 britain earmarked 37bn testing tracing 2020 2021 financial years though may well spend less shoddiness dogged elements campaign database cockup lost thousands results september 2020 remembereda basic problem contact tracing lot work theoretical assessments based analysing social networks experience asia suggest 30 contacts need identified index case digital tools lessen load resolve save lives campaign run vital strategies ngo developed locator taps credit data help tracers track people may caught virus specific index case amount work required remained enormousa much discussed alternative programmes use world ubiquitous tracking devices smartphones google apple worked together develop system enabled phones keep list occasions near another phone significant period time identity second phone someone tests positive sarscov2 asked send message phone used notify phones near within particular window time everything peertopeer neither big tech companies publichealth authorities get list contactsunfortunately builtin privacy makes hard assess technology efficacy british swiss studies suggest apps reduce spread enough make alsoran technology none successful contacttracing systems east asia relies things significant extenteven without good tracing self isolation tested positive helped slow spread disease time shortcomings initial testing technology reversetranscriptase polymerase chain reaction rtpcr became ever apparent conceptually elegant see diagram easy labs use despite familiarity reliability sensitivity real disadvantagesone needs labs carried efficiently big ones means samples may travel long way also means get held queues broad runs rtpcr tests three half hours average sample takes 15 hours process results rtpcr tests normally come days minutesanother problem though presence viral rna clearly shows person infected says little stand course disease rna detectible soon infection long disease run course publichealth terms needed test spots people actually infectious people cells noses throats actively churning virus particlesto look sarscov2 particles means looking distinctive protein components rna tells cells make detectible component spike protein studs particles outer membranes one basic rules modern biotechnology want find protein use antibodyantibodies large molecules come millions varieties sticks one target known antibody antigen one target handy technology called lateralflow testing makes use specificity sample placed one end porous membrane seeps along encounters line antibodies designed recognise sample urine antigen hormone found expectant women pregnancy test sample mucus swab antigen spike protein covid19 test cheap convenient one provide results half hour tests may pick people rtpcr might detect trace virus nose throat cells producing enough antigen test detect probably producing enough infectious eitherat beginning epidemic supply chain lateralflow tests says dr gabriel chris hand chairman abingdon health british contract manufacturer lateralflow tests says main bottleneck speciality membranes part every test kit come large reels 100 metres plus go automated equipment add biochemicals spraying low volumes says biochemicals bespoke antibodies generic bits bobs ready production processes place packaging sorted tests could churned millionnew technologies reaching market change dynamics test trace quantumdx one number companies developing automated pcrinabox systems provide results within couple hours jonathan halloran firm boss says british company relying testing system past 22 weeks testing 90 staff members every morning free decline week lunchtime brings news someone tested positive immediately sent home isolate things done fast fact rna detectable people infectious plus isolating basis early pcr test means one ever turns work infectious company claims lost single day infectiona combination local automated pcr lateralflow tests could basis ideal testing system one chance keeping ahead containing low level disease rather lagging behind one shooting antigen tests would used scan population new infections found would referred contact tracers contacts might infected could pcrtested find actually weregreat public health benefit index cases would wait isolated see fate virus immune systems store painfully aware next encounter wonders modern medical technology could hospital bed\",\n          \"decades scientists trying understand schizophrenia distressing disorder afflicts one hundred people destructive symptoms include hallucinations delusions muddled thoughts changes behaviour best drugs offer little ability target symptoms presently hope cure ultimate cause long mysteryschizophrenia known run families always thought genes might shed light origins disease genetics proved harder unpick anyone imagined 16 years human genome first sequenced scientists homed relevant genesfor five years researchers led group broad institute cambridge massachusetts collected 100000 dna samples around world looking regions human genome might harbouring variants increased risk schizophrenia found implicated hundred genes provided strong pointer towards portion genome associated infectious diseases many genes operate independently detective work revealed important source variation lay within particular gene known c4steven mccarroll harvard medical school colleagues involved study report nature week particular forms c4 gene led greater risk developing schizophrenia makes c4 hard identify comes many different forms intriguing function gene produces protein known complement factor helps tag pathogens removal body immune system brain however appears tied developmental process known synaptic pruning neurons continuously eliminated throughout childhood early adulthoodin words schizophrenics may producing much protein creates signal synaptic connections removed late adolescence brain undergoes widespread synaptic pruning particularly cerebral cortex also age schizophrenia symptoms tend begindavid goldstein director institute genomic medicine columbia university says persuaded research reckons important medical community examine study carefully region genome discussion hard work stuffed full genetic variation makes difficult point single gene say addsas whether family history schizophrenia ought genome sequenced assess susceptibility dr goldstein keen point work needs done risks different forms genes bring moreover factors implicated developing schizophrenia ranging taking psychoactive drugs exposed viruses wombsynaptic pruning crucial brain development researchers wonder mental disorders might tied aberrant activation regulation study bound initiate flurry similar work alzheimer disease one tied among things excessive destruction synapses already believed autistic individuals surplus synapses brain due slowdown normal brain pruning indeed onset autism often follows period burst synapse formation one big step forward schizophrenia research could enormous impact way neurological disorders viewed treated\"\n        ],\n        \"num_unique_values\": 961,\n        \"semantic_type\": \"\",\n        \"description\": \"\"\n      }\n    }\n  ]\n}"
            }
          },
          "metadata": {},
          "execution_count": 3
        }
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "grSc_-tmk3AJ"
      },
      "source": [
        "***quotations:***\n",
        "1. \"Some are born great. Some achieve greatness. Some have greatness thrust upon them.\" (W. Shakespeare, \"Twelfth Night\" Act II Scene 5)\n",
        "2. “So, naturalists observe, a flea/Hath smaller fleas that on him prey”.(Jonathan Swift...)\n",
        "3. “And these have smaller fleas to bite ’em/And so proceed ad infinitum” (Jonathan Swift...)\n",
        "4. \"...to hold infinity in the palm of a hand, and eternity in an hour.\" (William Blake ...)\n"
      ]
    },
    {
      "cell_type": "code",
      "source": [
        "model = Top2Vec.load(\"top2vec_model\")"
      ],
      "metadata": {
        "id": "zMZgYbUWAMnM"
      },
      "execution_count": 4,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [
        "topic_sizes = model.get_topic_sizes()\n",
        "topic_sizes"
      ],
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "BsBhliI2O4IB",
        "outputId": "bec9fbf0-312d-4c53-ee6f-e88826891674"
      },
      "execution_count": 5,
      "outputs": [
        {
          "output_type": "execute_result",
          "data": {
            "text/plain": [
              "(array([338, 259, 172,  77,  74,  42]), array([0, 1, 2, 3, 4, 5]))"
            ]
          },
          "metadata": {},
          "execution_count": 5
        }
      ]
    },
    {
      "cell_type": "markdown",
      "source": [
        "So, naturalists observe, a flea/Hath smaller fleas that on him prey."
      ],
      "metadata": {
        "id": "ESomTIV2CY-h"
      }
    },
    {
      "cell_type": "code",
      "execution_count": 6,
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "xD-aniRxzSUe",
        "outputId": "cd23231c-0479-4690-cc10-e3066433f44a"
      },
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "doc_id 780 Matches: see hybrids missing mechanism start 19th century jeanbaptiste lamarck french \u001b[38;5;1m\u001b[48;5;15mnaturalist thought\u001b[0m found recognised species mutable also proposed traits could inherited error\n",
            "doc_id 372 Matches: creature relatives seemed untroubled virusesas jonathan swift put muchmisquoted poem \u001b[38;5;1m\u001b[48;5;15mnaturalists observe\u001b[0m fleahath smaller fleas prey parasites words everywhere also usually abundant\n",
            "doc_id 524 Matches:  \u001b[38;5;1m\u001b[48;5;15m\u001b[0m \n",
            "doc_id 162 Matches: never climbed richard preston author several bestsellers tells story botanists \u001b[38;5;1m\u001b[48;5;15mnaturalists committed\u001b[0m preserving remaining fragment 2macre 810000 hectare redwood forest book revolves\n",
            "doc_id 168 Matches: dr aplin following work published 1934 henry smith williams american \u001b[38;5;1m\u001b[48;5;15mnaturalist noticed\u001b[0m put various coloured balls yarn garden almost always one one\n"
          ]
        }
      ],
      "source": [
        "# Получение информации о количестве документов в каждой теме и их индексах\n",
        "topic_sizes = model.get_topic_sizes()\n",
        "\n",
        "for topic_num, num_docs_in_topic in enumerate(topic_sizes[0]):\n",
        "    topic_index = topic_sizes[1][topic_num]  # Индекс текущей темы\n",
        "\n",
        "    documents, document_scores, document_ids = model.search_documents_by_topic(topic_num=topic_index, num_docs=num_docs_in_topic)\n",
        "\n",
        "    for text, score, doc_id in zip(documents, document_scores, document_ids):\n",
        "        matcher = Matcher(nlp.vocab)\n",
        "        pattern = [{'LEMMA': \"naturalist\"}, {'POS': \"VERB\"}] # Шаблон для поиска искомой фразы\n",
        "        matcher.add(\"naturalist_PATTERN\", [pattern])\n",
        "\n",
        "        doc = nlp(str(text))\n",
        "        matches = matcher(doc)\n",
        "\n",
        "        if len(matches) > 0:\n",
        "            result = [doc[start - 10:end + 10].text for match_id, start, end in matches][0].split()\n",
        "            highlighted_phrase = ' '.join(result[10:12]) # Срез искомой фразы\n",
        "            highlighted_text = ' '.join(result[:10]) + f' {Fore.RED}{Back.WHITE}{highlighted_phrase}{Style.RESET} ' + ' '.join(result[-10:])\n",
        "            print(\"doc_id\", doc_id, \"Matches:\", highlighted_text)"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": 7,
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/",
          "height": 143
        },
        "id": "wj1-_r0pGcTI",
        "outputId": "cd56693c-9cbd-49a1-9bd3-1b5bc3c8f090"
      },
      "outputs": [
        {
          "output_type": "execute_result",
          "data": {
            "text/plain": [
              "'WHAT is the commonest living thing on Earth? Until now, those in the know would probably have answered Pelagibacter ubique, the most successful member of a group of bacteria, called SAR11, that jointly constitute about a third of the single-celled organisms in the ocean. But this is not P. ubique’s only claim to fame, for unlike almost every other known cellular creature, it and its relatives have seemed to be untroubled by viruses.As Jonathan Swift put it in a much-misquoted poem, “So, naturalists observe, a flea/Hath smaller fleas that on him prey”. Parasites, in other words, are everywhere. They are also, usually, more abundant than their hosts. An astute observer might therefore have suspected that the actual most-common species on Earth would be a “flea” that parasitised P. ubique, rather than the bacterium itself. The absence of such fleas (in the form of viruses called bacteriophages, that attack bacteria) has puzzled virologists since 1990, when the SAR11 group was identified. Some thought the advantage this absence conferred explained the group’s abundance. But no. As they report in this week’s Nature, Stephen Giovannoni of Oregon State University and his colleagues have discovered the elusive phages. Swift’s wisdom, it seems, still holds good.Tracking down a particular virus in the ocean makes finding a needle in a haystack look a trivial task. A litre of seawater has billions of viruses in it. Modern genetic techniques can obtain DNA sequences from these viruses, but that cannot tie a particular virus to a particular host.To do so, Dr Giovannoni (pictured) borrowed a technique from homeopathy: he diluted some seawater to such an extent that, statistically speaking, he expected a 100-microlitre-sized aliquot to contain only one or two viruses. The difference between his approach and a homeopath’s was that what homeopathy dilutes almost to nothing are chemicals, and thus cannot breed. A virus can, given a suitable host. So he mixed each of several hundred aliquots into tubes of water containing P. ubique. Then he waited.After 60 hours, he looked to see what had happened. In most cases the bacteria had thrived. In a few, though, they had been killed by what looked like viral infection. It was these samples that he ran through the DNA-sequencing machine, in the knowledge that the only viral DNA present would be from whatever it was had killed the bacteria.His reward was to find not one, but four viruses that parasitise P. ubique. He then compared their DNA with databases of DNA found in seawater from around the world, to find out how abundant each is. The upshot was that a virus dubbed HTVC010P was the commonest. It thus displaces its host as the likely winner of the most-common-living-thing prize.That does depend, of course, on your definition of “living thing”. Some biologists count viruses as organisms. Some do not. The reason is that a virus relies for its growth and reproduction on the metabolic processes of the cell it infects. This means viruses themselves are hard to parasitise, since they do no work on which another organism can free-ride. Which is why the next two lines of Swift’s poem, “And these have smaller fleas to bite ’em/And so proceed ad infinitum”, are wrong—and why, because HTVC010P itself can have no parasites, it probably really is the commonest organism on the planet.'"
            ],
            "application/vnd.google.colaboratory.intrinsic+json": {
              "type": "string"
            }
          },
          "metadata": {},
          "execution_count": 7
        }
      ],
      "source": [
        "idx = 372\n",
        "res_df = df.iloc[idx, :]\n",
        "res_df['text']\n"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": 8,
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "HJF9xplbGnBp",
        "outputId": "e18b93d5-b741-4f41-e7f8-c54aa4fffc07"
      },
      "outputs": [
        {
          "output_type": "execute_result",
          "data": {
            "text/plain": [
              "text                  WHAT is the commonest living thing on Earth? U...\n",
              "title                                                       Flea market\n",
              "date                                                      Feb 16th 2013\n",
              "link                  https://www.economist.com/science-and-technolo...\n",
              "cleaned_text          what is the commonest living thing on earth un...\n",
              "fully_cleaned_text    commonest living thing earth know would probab...\n",
              "Name: 372, dtype: object"
            ]
          },
          "metadata": {},
          "execution_count": 8
        }
      ],
      "source": [
        "res_df"
      ]
    }
  ]
}